{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0c35f698",
   "metadata": {},
   "source": [
    "# Package Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "608b95ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np # Fundamental package for scientific computing with Python\n",
    "import pandas as pd # Data manipulation package, especially for tabular data, most downladed public python package\n",
    "import re # Regular expression operations (string matching and manipulation)\n",
    "\n",
    "from typing import Any, Optional, Iterable, List\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cb34405",
   "metadata": {},
   "source": [
    "# Filepaths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "96e929f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp_biomarkers = \"Data/LS Biomarkers Timeseries.csv\" # Read in the biomarker time series data\n",
    "fp_biomarkersLong = \"Data/LLM_Biomarkers_Long.csv\" # Write out the long format data\n",
    "fp_biomarkersWide = \"Data/LLM_Biomarkers_Wide.csv\" # Write out the wide format data\n",
    "\n",
    "fp_microbiome = \"Data/LS Gut Microbiome.csv\" # Read in the microbiome data\n",
    "fp_microbiomeLong = \"Data/LLM_Gut_Microbiome_Long.csv\" # Write out the long format data\n",
    "fp_microbiomeWide = \"Data/LLM_Gut_Microbiome_Wide.csv\" # Write out the wide format data\n",
    "fp_microbiomeSummary = \"Data/LLM_Gut_Microbiome_Summary.csv\" # Write out the summary data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b248d661",
   "metadata": {},
   "source": [
    "# Biomarker Time Series Data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "ab9abea9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date collected</th>\n",
       "      <th>Units</th>\n",
       "      <th>Good Range</th>\n",
       "      <th>5/1/1993</th>\n",
       "      <th>11/27/1996</th>\n",
       "      <th>5/8/1997</th>\n",
       "      <th>4/15/1998</th>\n",
       "      <th>4/1/1999</th>\n",
       "      <th>10/6/1999</th>\n",
       "      <th>5/7/2001</th>\n",
       "      <th>...</th>\n",
       "      <th>2/5/2025</th>\n",
       "      <th>2/12/2025</th>\n",
       "      <th>2/28/2025</th>\n",
       "      <th>5/20/2025</th>\n",
       "      <th>6/13/2025</th>\n",
       "      <th>Low</th>\n",
       "      <th>High</th>\n",
       "      <th>Average</th>\n",
       "      <th>Number</th>\n",
       "      <th>Unnamed: 341</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>YFH/UCSD</td>\n",
       "      <td>UCSD</td>\n",
       "      <td>UCSD</td>\n",
       "      <td>UCSD</td>\n",
       "      <td>UCSD</td>\n",
       "      <td>of tests since '07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Weight</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>202.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>...</td>\n",
       "      <td>178</td>\n",
       "      <td>179</td>\n",
       "      <td>179.6</td>\n",
       "      <td>182.5</td>\n",
       "      <td>180.8</td>\n",
       "      <td>166.1</td>\n",
       "      <td>202.0</td>\n",
       "      <td>185.6</td>\n",
       "      <td>322.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Blood Tests</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>Butyrate</td>\n",
       "      <td>mg/mL</td>\n",
       "      <td>0.8-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.3</td>\n",
       "      <td>4.3</td>\n",
       "      <td>1.3</td>\n",
       "      <td>184.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>238 rows × 342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Date collected  Units Good Range  5/1/1993  11/27/1996  5/8/1997  \\\n",
       "0              NaN    NaN        NaN       NaN         NaN       NaN   \n",
       "1              NaN    NaN        NaN       NaN         NaN       NaN   \n",
       "2           Weight    NaN        NaN       NaN         NaN       NaN   \n",
       "3              NaN    NaN        NaN       NaN         NaN       NaN   \n",
       "4      Blood Tests    NaN        NaN       NaN         NaN       NaN   \n",
       "..             ...    ...        ...       ...         ...       ...   \n",
       "233       Butyrate  mg/mL      0.8-4       NaN         NaN       NaN   \n",
       "234            NaN    NaN        NaN       NaN         NaN       NaN   \n",
       "235            NaN    NaN        NaN       NaN         NaN       NaN   \n",
       "236            NaN    NaN        NaN       NaN         NaN       NaN   \n",
       "237            NaN    NaN        NaN       NaN         NaN       NaN   \n",
       "\n",
       "     4/15/1998  4/1/1999  10/6/1999  5/7/2001  ...  2/5/2025  2/12/2025  \\\n",
       "0          NaN       NaN        NaN       NaN  ...  YFH/UCSD       UCSD   \n",
       "1          NaN       NaN        NaN       NaN  ...       NaN        NaN   \n",
       "2          NaN       NaN      202.0     201.0  ...       178        179   \n",
       "3          NaN       NaN        NaN       NaN  ...       NaN        NaN   \n",
       "4          NaN       NaN        NaN       NaN  ...       NaN        NaN   \n",
       "..         ...       ...        ...       ...  ...       ...        ...   \n",
       "233        NaN       NaN        NaN       NaN  ...         2        NaN   \n",
       "234        NaN       NaN        NaN       NaN  ...       NaN        NaN   \n",
       "235        NaN       NaN        NaN       NaN  ...       NaN        NaN   \n",
       "236        NaN       NaN        NaN       NaN  ...       NaN        NaN   \n",
       "237        NaN       NaN        NaN       NaN  ...       NaN        NaN   \n",
       "\n",
       "     2/28/2025 5/20/2025  6/13/2025                 Low   High Average Number  \\\n",
       "0         UCSD      UCSD       UCSD  of tests since '07    NaN     NaN    NaN   \n",
       "1          NaN       NaN        NaN                 NaN    NaN     NaN    NaN   \n",
       "2        179.6     182.5      180.8               166.1  202.0   185.6  322.0   \n",
       "3          NaN       NaN        NaN                 NaN    NaN     NaN    NaN   \n",
       "4          NaN       NaN        NaN                 NaN    NaN     NaN    NaN   \n",
       "..         ...       ...        ...                 ...    ...     ...    ...   \n",
       "233        NaN       NaN        NaN                 0.3    4.3     1.3  184.0   \n",
       "234        NaN       NaN        NaN                 NaN    NaN     NaN    NaN   \n",
       "235        NaN       NaN        NaN                 NaN    NaN     NaN    NaN   \n",
       "236        NaN       NaN        NaN                 NaN    NaN     NaN    NaN   \n",
       "237        NaN       NaN        NaN                 NaN    NaN     NaN    NaN   \n",
       "\n",
       "    Unnamed: 341  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "..           ...  \n",
       "233          NaN  \n",
       "234          NaN  \n",
       "235          NaN  \n",
       "236          NaN  \n",
       "237          NaN  \n",
       "\n",
       "[238 rows x 342 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_biomarkers = pd.read_csv(fp_biomarkers)\n",
    "df_biomarkers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1b6bbc9",
   "metadata": {},
   "source": [
    "## Create Long Format"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbdae351",
   "metadata": {},
   "source": [
    "### Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "680f2963",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_blank(x: Any) -> bool:\n",
    "    \"\"\" Return True if `x` is effectively empty: None, NaN, or a string of only whitespace.\n",
    "    Parameters-\n",
    "    x : Any, Value to check\n",
    "    Returns-\n",
    "    bool: True if `x` is None, NaN, or a blank/whitespace-only string; False otherwise.\"\"\"\n",
    "    \n",
    "    return (x is None) or (pd.isna(x)) or (str(x).strip() == \"\")\n",
    "\n",
    "def find_col(name_like: str, default: Optional[str] = None) -> Optional[str]:\n",
    "    \"\"\" Return the FIRST column in global `cols` whose (lowercased, trimmed) text\n",
    "    contains the (lowercased, trimmed) substring `name_like`.\n",
    "    Parameters-\n",
    "    name_like : str, Substring to look for in each column name.\n",
    "    default : Optional[str], default None, Value to return if no match is found.\"\"\"\n",
    "    \n",
    "    for c in cols:\n",
    "        if name_like in str(c).strip().lower():\n",
    "            return c\n",
    "    return default\n",
    "\n",
    "def find_exact(label: str) -> Optional[str]:\n",
    "    \"\"\" Return an EXACT matching column from global `cols`. If not found,\n",
    "    return a case-insensitive match on the whole label (after trimming).\n",
    "    Parameters-\n",
    "    label : str, Column label to match.\n",
    "    Returns-\n",
    "    Optional[str]: Matching column name from `cols`, or None if no match.\"\"\"\n",
    "    \n",
    "    for c in cols:\n",
    "        if c == label:\n",
    "            return c\n",
    "    low = label.lower()\n",
    "    for c in cols:\n",
    "        if str(c).strip().lower() == low:\n",
    "            return c\n",
    "    return None\n",
    "\n",
    "def uniq_order(seq):\n",
    "    \"\"\" Deduplicate while preserving first-seen order.\n",
    "    - Converts items to trimmed strings.\n",
    "    - Skips None/blank items.\n",
    "    Parameters-\n",
    "    seq : Iterable[Any], Sequence of values.\n",
    "    Returns-\n",
    "    List[str]: List of unique, trimmed, non-blank strings in their first-seen order. \"\"\"\n",
    "    \n",
    "    seen = set()\n",
    "    out = []\n",
    "    for x in seq:\n",
    "        if x is None: \n",
    "            continue\n",
    "        s = str(x).strip()\n",
    "        if not s or s in seen:\n",
    "            continue\n",
    "        seen.add(s)\n",
    "        out.append(s)\n",
    "    return out\n",
    "\n",
    "def clean_label(x):\n",
    "    \"\"\" Normalize a label/string:\n",
    "    - Return None if blank (None/NaN/whitespace-only).\n",
    "    - Replace non-breaking spaces (\\\\u00A0) with regular spaces.\n",
    "    - Trim outer whitespace.\n",
    "\n",
    "    Parameters-\n",
    "    x : Any, Value to clean.\n",
    "    Returns-\n",
    "    Optional[str]: Cleaned string, or None if input is blank.\"\"\"\n",
    "    if is_blank(x):\n",
    "        return None\n",
    "    return str(x).replace(\"\\u00a0\", \" \").strip()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c01fdef8",
   "metadata": {},
   "source": [
    "### Mapping Dictionaries for Columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "0288ecd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping For biomarkeers to panels\n",
    "dict_biomarkerToPanels = {\n",
    "                            # Vitals\n",
    "                            \"Weight\": \"Vitals\",\n",
    "\n",
    "                            # LIPID Panel\n",
    "                            \"237\": \"LIPID Panel\",\n",
    "                            'HDL (High Density Lipoprotein) \"good\"': \"LIPID Panel\",\n",
    "                            \"VLDL Cholesterol\": \"LIPID Panel\",\n",
    "                            \"IDL Chloresterol\": \"LIPID Panel\",\n",
    "                            \"Sum Total Cholesterol\": \"LIPID Panel\",\n",
    "                            \"Triglycerides\": \"LIPID Panel\",\n",
    "                            \"UCSD LDL Particle # by NMR\": \"LIPID Panel\",\n",
    "                            \"UCSD HDL Particle #\": \"LIPID Panel\",\n",
    "                            \"UCSD Large VLDL Particle #\": \"LIPID Panel\",\n",
    "                            \"UCSD Small LDL Particle #\": \"LIPID Panel\",\n",
    "                            \"UCSD Large HDL Particle #\": \"LIPID Panel\",\n",
    "                            \"UCSD VLDL Particle Size\": \"LIPID Panel\",\n",
    "                            \"UCSD LDL Particle Size\": \"LIPID Panel\",\n",
    "                            \"UCSD HDL Particle Size\": \"LIPID Panel\",\n",
    "                            \"Real-LDL Size Pattern\": \"LIPID Panel\",\n",
    "                            \"TG/HDL\": \"LIPID Panel\",\n",
    "                            \"Chol/HDL\": \"LIPID Panel\",\n",
    "                            \"Lp(a) Chloresterol\": \"LIPID Panel\",\n",
    "                            \"APO A1\": \"LIPID Panel\",\n",
    "                            \"APOLIPOPROTEIN-(B100)\": \"LIPID Panel\",\n",
    "                            \"Ratio APO B100/AI\": \"LIPID Panel\",\n",
    "\n",
    "                            # Comprehensive Metabolic Panel – Electrolytes\n",
    "                            \"Sodium\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Potassium\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Chloride\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Calcium\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Phosporus\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Magnesium\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Zinc\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Tot CO2\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Bicarbonate\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Anion Gap\": \"Comprehensive Metabolic Panel\",\n",
    "                            \n",
    "                            # Comprehensive Metabolic Panel – remainder\n",
    "                            \"Total Protein\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Albumin\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Globulin\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"A/G ratio\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"GGTP\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"LDH\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"ALT\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"AST\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"AST/ALT ratio\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Alkaline phosphatase\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Total Bilirubin\": \"Comprehensive Metabolic Panel\",\n",
    "                            \"Direct Bilirubin\": \"Comprehensive Metabolic Panel\",\n",
    "\n",
    "                            # Kidney\n",
    "                            \"Blood Urea Nitrogen (BUN)\": \"Kidney\",\n",
    "                            \"Creatinine\": \"Kidney\",\n",
    "                            \"BUN/Creatinine Ratio\": \"Kidney\",\n",
    "                            \"Uric Acid\": \"Kidney\",\n",
    "\n",
    "                            # Glucose/Insulin Panel\n",
    "                            \"Fasting Glucose\": \"Glucose/Insulin Panel\",\n",
    "                            \"Insulin\": \"Glucose/Insulin Panel\",\n",
    "                            \"Hemoglobin A1C\": \"Glucose/Insulin Panel\",\n",
    "                            \"HOMA-IR (insulin resistance)\": \"Glucose/Insulin Panel\",\n",
    "\n",
    "                            # Hormone Panel\n",
    "                            \"Free Testosterone\": \"Hormone Panel\",\n",
    "                            \"Testosterone % Free\": \"Hormone Panel\",\n",
    "                            \"Total Testosterone\": \"Hormone Panel\",\n",
    "                            \"Sex Hormone Binding Globulin (SHBG)\": \"Hormone Panel\",\n",
    "                            \"Estradiol\": \"Hormone Panel\",\n",
    "                            \"DHEA-blood\": \"Hormone Panel\",\n",
    "                            \"Cortisol Random\": \"Hormone Panel\",\n",
    "                            \"TSH\": \"Hormone Panel\",\n",
    "                            \"Total T3\": \"Hormone Panel\",\n",
    "                            \"Thyroxine (T4)\": \"Hormone Panel\",\n",
    "\n",
    "                            # Cancer Panel\n",
    "                            \"PSA Screen\": \"Cancer Panel\",\n",
    "                            \"CA-19-9 (Pancreatic Cancer)\": \"Cancer Panel\",\n",
    "                            \"CEA\": \"Cancer Panel\",\n",
    "\n",
    "                            # Iron Panel\n",
    "                            \"Iron\": \"Iron Panel\",\n",
    "                            \"Ferritin\": \"Iron Panel\",\n",
    "                            \"Fibrinogen\": \"Iron Panel\",\n",
    "                            \"TIBC\": \"Iron Panel\",\n",
    "                            \"UIBC\": \"Iron Panel\",\n",
    "                            \"Iron saturation\": \"Iron Panel\",\n",
    "\n",
    "                            # CBC – Red blood cell details\n",
    "                            \"RBC-\": \"CBC: Red Blood Cell Details\",\n",
    "                            \"HGB-\": \"CBC: Red Blood Cell Details\",\n",
    "                            \"HCT\": \"CBC: Red Blood Cell Details\",\n",
    "                            \"PLT-\": \"CBC: Red Blood Cell Details\",\n",
    "                            \"MCV-\": \"CBC: Red Blood Cell Details\",\n",
    "                            \"MCH\": \"CBC: Red Blood Cell Details\",\n",
    "                            \"MCHC-\": \"CBC: Red Blood Cell Details\",\n",
    "                            \"RDW\": \"CBC: Red Blood Cell Details\",\n",
    "                            \"MPV\": \"CBC: Red Blood Cell Details\",\n",
    "\n",
    "                            # CBC – White blood cell details\n",
    "                            \"WBC-\": \"CBC: White Blood Cell Details\",\n",
    "                            \"NEU %\": \"CBC: White Blood Cell Details\",\n",
    "                            \"LYM %\": \"CBC: White Blood Cell Details\",\n",
    "                            \"EOS %\": \"CBC: White Blood Cell Details\",\n",
    "                            \"MONO %\": \"CBC: White Blood Cell Details\",\n",
    "                            \"BASO %\": \"CBC: White Blood Cell Details\",\n",
    "                            \"Neutrophil Count\": \"CBC: White Blood Cell Details\",\n",
    "                            \"Lymphocyte Count\": \"CBC: White Blood Cell Details\",\n",
    "                            \"Monocyte Count\": \"CBC: White Blood Cell Details\",\n",
    "                            \"Esoinophil Count\": \"CBC: White Blood Cell Details\",\n",
    "                            \"Basophil Count\": \"CBC: White Blood Cell Details\",\n",
    "\n",
    "                            # Inflammation Panel\n",
    "                            \"High Sensitivity CRP (mg/L)\": \"Inflammation Panel\",\n",
    "                            \"Lp-PLA2\": \"Inflammation Panel\",\n",
    "                            \"IgG\": \"Inflammation Panel\",\n",
    "                            \"IgA\": \"Inflammation Panel\",\n",
    "                            \"IgM\": \"Inflammation Panel\",\n",
    "\n",
    "                            # Vitamins, Antioxidants, Amino Acid Panel\n",
    "                            \"Homocysteine\": \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "                            \"CoQ10\": \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "                            \"Selenium\": \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "                            \"Vitamin E Alpha-Tocopherol\": \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "                            \"Vitamin E Gamma-Tocopherol\": \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "                            \"Alpha Lipoic Acid\": \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "                            \"Vitamin D, 25-Hydroxy\": \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "                            \"Vitamin B12\": \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "                            \"Niacin (Vit B3)\": \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "\n",
    "                            # Stool tests\n",
    "                            \"pH\": \"Stool tests\",\n",
    "\n",
    "                            # Cultured Bacteria Growth, Expected/Beneficial flora\n",
    "                            \"Bifidobacerium family\": \"Expected/Beneficial flora\",\n",
    "                            \"E. coli spp.\": \"Expected/Beneficial flora\",\n",
    "                            \"Lactobacillus spp\": \"Expected/Beneficial flora\",\n",
    "                            \"Enterococcus spp.\": \"Expected/Beneficial flora\",\n",
    "                            \"Bacteroides fraglis group\": \"Expected/Beneficial flora\",\n",
    "                            \"Clostridium spp\": \"Expected/Beneficial flora\",\n",
    "                            \"Bacteroides family\": \"Expected/Beneficial flora\",\n",
    "\n",
    "                            # Cultured Bacteria Growth, Commensal (Imbalanced) flora\n",
    "                            \"Alpha hemolytic strep\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Hemolytic Escherichia coli\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Pseudomonas chlororaphis group\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Gamma Hemolytic Streptococcus\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Hafnia alvei\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Mucoid Escherichia coli\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Beta strep, group B, strep. Agalactiae\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Klebsiella oxytoca\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Bacillus spp, pumilus\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Comamonas testosteroni\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Streptococcus anginosus\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Cellulosimicrobium cellulans/funkeil/marinum\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Lactococcus lactis\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Staphylococcus aureus\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Streptococcus salivarius\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Streptococcus parasanguinis\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Pseudomonas mendocina \": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Actinomyces turicensis\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Lysinibacillus spp\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Streptococcus mutans\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Streptococcus australis\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Enterobacter asburiae\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Pseudomonas stutzeri\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Kocuria spp, salsicia, rhizophila, kristinae\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Rothia dentocariosa/mucliaginosa\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Pantoea spp\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Pseudomonas citronellolis\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Streptococcus mitis/oralis/cristatus\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Staphylococcus haemolyticus\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Serratia marcescens\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Staphylococcus lugdunensis\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Staphylococcus epidermidis\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Actinomyces neuii\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Streptomyces\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Bacillus licheniformis\": \"Commensal (Imbalanced) flora\",\n",
    "                            \"Corynebacterium falsenii\": \"Commensal (Imbalanced) flora\",\n",
    "\n",
    "                            # Dysbiotic bacteria\n",
    "                            \"Morganellla morganii\": \"Dysbiotic bacteria\",\n",
    "                            \"Providencia stuartii\": \"Dysbiotic bacteria\",\n",
    "                            \"Klebsiella pneumonia\": \"Dysbiotic bacteria\",\n",
    "                            \"Providencia rettgeri\": \"Dysbiotic bacteria\",\n",
    "                            \"Enterobacter cloacae\": \"Dysbiotic bacteria\",\n",
    "                            \"Pseudomonas aeruginosa\": \"Dysbiotic bacteria\",\n",
    "                            \"Bacillus cereus\": \"Dysbiotic bacteria\",\n",
    "                            \"Citrobacter freundii complex\": \"Dysbiotic bacteria\",\n",
    "                            \"Arcobacter butzieri\": \"Dysbiotic bacteria\",\n",
    "\n",
    "                            # Yeast\n",
    "                            \"Candida parapsilosis\": \"Yeast\",\n",
    "                            \"Rhodotorula glutinis/mucilaginosa\": \"Yeast\",\n",
    "                            \"Candida krusei/inconspicua/lambica\": \"Yeast\",\n",
    "                            \"Saccharomyces cerevisiae/boulardi\": \"Yeast\",\n",
    "                            \"Geotrichum spp.\": \"Yeast\",\n",
    "                            \"Candida pararugosa\": \"Yeast\",\n",
    "                            \"Candida lusitaniae\": \"Yeast\",\n",
    "                            \"Candida zeylanoides\": \"Yeast\",\n",
    "                            \"Trichosporon asahii\": \"Yeast\",\n",
    "                            \"Trichosporon ovoides\": \"Yeast\",\n",
    "                            \"Trichosporon montevideense\": \"Yeast\",\n",
    "                            \"Exophiala dermatitidis\": \"Yeast\",\n",
    "                            \"Cryptococcus spp, not neoformans\": \"Yeast\",\n",
    "                            \"Cryptococcus diffuens\": \"Yeast\",\n",
    "                            \"Candida orthopsilosis\": \"Yeast\",\n",
    "\n",
    "                            # Gut Inflammation\n",
    "                            \"Stool Calprotectin UCSD\": \"Gut Inflammation\",\n",
    "                            \"Stool Calprotectin YFH\": \"Gut Inflammation\",\n",
    "                            \"Stool Lysozyme\": \"Gut Inflammation\",\n",
    "                            \"Stool Lactoferrin\": \"Gut Inflammation\",\n",
    "                            \"Stool SIgA\": \"Gut Inflammation\",\n",
    "\n",
    "                             # Short Chain Fatty Acids\n",
    "                            \"% SCFA Acetate\": \"Short Chain Fatty Acids\",\n",
    "                            \"% SCFA Propionate\": \"Short Chain Fatty Acids\",\n",
    "                            \"% SCFA Valerate\": \"Short Chain Fatty Acids\",\n",
    "                            \"% SCFA Butyrate\": \"Short Chain Fatty Acids\",\n",
    "                            \"Total SCFA\": \"Short Chain Fatty Acids\",\n",
    "                            \"Butyrate\": \"Short Chain Fatty Acids\"\n",
    "\n",
    "}\n",
    "\n",
    "# Mapping for units to biomarkers\n",
    "dict_unitsToBiomarkers = {\n",
    "                            # Vitals\n",
    "                            \"Weight\": \"lbs\",\n",
    "\n",
    "                            # LIPID Panel\n",
    "                            \"237\": \"mg/dL\",\n",
    "                            'HDL (High Density Lipoprotein) \"good\"': \"mg/dL\",\n",
    "                            \"VLDL Cholesterol\": \"mg/dL\",\n",
    "                            \"IDL Chloresterol\": None,\n",
    "                            \"Sum Total Cholesterol\": \"mg/dL\",\n",
    "                            \"Triglycerides\": \"mg/dL\",\n",
    "                            \"UCSD LDL Particle # by NMR\": \"nmol/L\",\n",
    "                            \"UCSD HDL Particle #\": \"μmol/L\",\n",
    "                            \"UCSD Large VLDL Particle #\": \"nmol/L\",\n",
    "                            \"UCSD Small LDL Particle #\": \"nmol/L\",\n",
    "                            \"UCSD Large HDL Particle #\": \"μmol/L\",\n",
    "                            \"UCSD VLDL Particle Size\": \"nm\",\n",
    "                            \"UCSD LDL Particle Size\": \"nm\",\n",
    "                            \"UCSD HDL Particle Size\": \"nm\",\n",
    "                            \"Real-LDL Size Pattern\": None,\n",
    "                            \"TG/HDL\": None,\n",
    "                            \"Chol/HDL\": None,\n",
    "                            \"Lp(a) Chloresterol\": None,\n",
    "                            \"APO A1\": \"mg/dL\",\n",
    "                            \"APOLIPOPROTEIN-(B100)\": \"mg/dL\",\n",
    "                            \"Ratio APO B100/AI\": None,\n",
    "\n",
    "                            # Comprehensive Metabolic Panel – Electrolytes\n",
    "                            \"Sodium\": \"mmol/L\",\n",
    "                            \"Potassium\": \"mmol/L\",\n",
    "                            \"Chloride\": \"mmol/L\",\n",
    "                            \"Calcium\": \"mg/dL\",\n",
    "                            \"Phosporus\": \"mg/dL\",\n",
    "                            \"Magnesium\": \"mg/dL\",\n",
    "                            \"Zinc\": \"µg/dL\",\n",
    "                            \"Tot CO2\": None,\n",
    "                            \"Bicarbonate\": \"mmol/L\",\n",
    "                            \"Anion Gap\": \"mmol/L\",\n",
    "\n",
    "                            # Comprehensive Metabolic Panel – remainder\n",
    "                            \"Total Protein\": \"g/dL\",\n",
    "                            \"Albumin\": \"g/dL\",\n",
    "                            \"Globulin\": None,\n",
    "                            \"A/G ratio\": None,\n",
    "                            \"GGTP\": \"U/L\",\n",
    "                            \"LDH\": \"U/L\",\n",
    "                            \"ALT\": \"U/L\",\n",
    "                            \"AST\": \"U/L\",\n",
    "                            \"AST/ALT ratio\": None,\n",
    "                            \"Alkaline phosphatase\": \"U/L\",\n",
    "                            \"Total Bilirubin\": \"mg/dL\",\n",
    "                            \"Direct Bilirubin\": None,\n",
    "\n",
    "                            # Kidney\n",
    "                            \"Blood Urea Nitrogen (BUN)\": \"mg/dL\",\n",
    "                            \"Creatinine\": \"mg/dL\",\n",
    "                            \"BUN/Creatinine Ratio\": None,\n",
    "                            \"Uric Acid\": None,\n",
    "\n",
    "                            # Glucose/Insulin Panel\n",
    "                            \"Fasting Glucose\": \"mg/dL\",\n",
    "                            \"Insulin\": \"µIU/mL\",\n",
    "                            \"Hemoglobin A1C\": \"%\",\n",
    "                            \"HOMA-IR (insulin resistance)\": None,\n",
    "\n",
    "                            # Hormone Panel\n",
    "                            \"Free Testosterone\": \"pg/mL\",\n",
    "                            \"Testosterone % Free\": \"%\",\n",
    "                            \"Total Testosterone\": \"ng/mL\",\n",
    "                            \"Sex Hormone Binding Globulin (SHBG)\": \"nmol/L\",\n",
    "                            \"Estradiol\": \"pg/mL\",\n",
    "                            \"DHEA-blood\": \"µg/dL\",\n",
    "                            \"Cortisol Random\": None,\n",
    "                            \"TSH\": \"µIU/mL\",\n",
    "                            \"Total T3\": \"ng/mL\",\n",
    "                            \"Thyroxine (T4)\": \"µg/dL\",\n",
    "\n",
    "                            # Cancer Panel\n",
    "                            \"PSA Screen\": None,\n",
    "                            \"CA-19-9 (Pancreatic Cancer)\": \"U/mL\",\n",
    "                            \"CEA\": \"ng/mL\",\n",
    "\n",
    "                            # Iron Panel\n",
    "                            \"Iron\": \"µg/dL\",\n",
    "                            \"Ferritin\": \"ng/mL\",\n",
    "                            \"Fibrinogen\": None,\n",
    "                            \"TIBC\": \"µg/dL\",\n",
    "                            \"UIBC\": \"µg/dL\",\n",
    "                            \"Iron saturation\": \"%\",\n",
    "\n",
    "                            # Complete Blood Count (CBC) Panel – Red Blood Cell Details\n",
    "                            \"RBC-\": \"10**6/mL\",\n",
    "                            \"HGB-\": \"g/dL\",\n",
    "                            \"HCT\": \"%\",\n",
    "                            \"PLT-\": \"10**3/mL\",\n",
    "                            \"MCV-\": \"fL\",\n",
    "                            \"MCH\": \"pg/cell\",\n",
    "                            \"MCHC-\": \"g/dL\",\n",
    "                            \"RDW\": \"%\",\n",
    "                            \"MPV\": \"fL\",\n",
    "\n",
    "                            # Complete Blood Count (CBC) Panel  – White Blood Cell Details\n",
    "                            \"WBC-\": \"10**3/mL\",\n",
    "                            \"NEU %\": \"%\",\n",
    "                            \"LYM %\": \"%\",\n",
    "                            \"EOS %\": \"%\",\n",
    "                            \"MONO %\": \"%\",\n",
    "                            \"BASO %\": \"%\",\n",
    "                            \"Neutrophil Count\": \"10**3/mL\",\n",
    "                            \"Lymphocyte Count\": \"10**3/mL\",\n",
    "                            \"Monocyte Count\": \"10**3/mL\",\n",
    "                            \"Esoinophil Count\": \"10**3/mL\",\n",
    "                            \"Basophil Count\": \"10**3/mL\",\n",
    "\n",
    "                            # Inflammation Panel\n",
    "                            \"High Sensitivity CRP (mg/L)\": \"mg/L\",\n",
    "                            \"Lp-PLA2\": None,\n",
    "                            \"IgG\": None,\n",
    "                            \"IgA\": None,\n",
    "                            \"IgM\": None,\n",
    "\n",
    "                            # Vitamins, Antioxidants, Amino Acid Panel\n",
    "                            \"Homocysteine\": \"µmol/L\",\n",
    "                            \"CoQ10\": None,\n",
    "                            \"Selenium\": None,\n",
    "                            \"Vitamin E Alpha-Tocopherol\": \"mg/L\",\n",
    "                            \"Vitamin E Gamma-Tocopherol\": \"mg/L\",\n",
    "                            \"Alpha Lipoic Acid\": None,\n",
    "                            \"Vitamin D, 25-Hydroxy\": \"ng/mL\",\n",
    "                            \"Vitamin B12\": \"pg/mL\",\n",
    "                            \"Niacin (Vit B3)\": None,\n",
    "\n",
    "                            # Stool Tests\n",
    "                            \"pH\": None,\n",
    "\n",
    "                            # Cultured Bacteria Growth - Expected/Beneficial flora\n",
    "                            \"Bifidobacerium family\": None,\n",
    "                            \"E. coli spp.\": None,\n",
    "                            \"Lactobacillus spp\": None,\n",
    "                            \"Enterococcus spp.\": None,\n",
    "                            \"Bacteroides fraglis group\": None,\n",
    "                            \"Clostridium spp\": None,\n",
    "                            \"Bacteroides family\": None,\n",
    "\n",
    "                            # Cultured Bacteria Growth - Commensal (Imbalanced) flora \n",
    "                            \"Alpha hemolytic strep\": None,\n",
    "                            \"Hemolytic Escherichia coli\": None,\n",
    "                            \"Pseudomonas chlororaphis group\": None,\n",
    "                            \"Gamma Hemolytic Streptococcus\": None,\n",
    "                            \"Hafnia alvei\": None,\n",
    "                            \"Mucoid Escherichia coli\": None,\n",
    "                            \"Beta strep, group B, strep. Agalactiae\": None,\n",
    "                            \"Klebsiella oxytoca\": None,\n",
    "                            \"Bacillus spp, pumilus\": None,\n",
    "                            \"Comamonas testosteroni\": None,\n",
    "                            \"Streptococcus anginosus\": None,\n",
    "                            \"Cellulosimicrobium cellulans/funkeil/marinum\": None,\n",
    "                            \"Lactococcus lactis\": None,\n",
    "                            \"Staphylococcus aureus\": None,\n",
    "                            \"Streptococcus salivarius\": None,\n",
    "                            \"Streptococcus parasanguinis\": None,\n",
    "                            \"Pseudomonas mendocina \": None,\n",
    "                            \"Actinomyces turicensis\": None,\n",
    "                            \"Lysinibacillus spp\": None,\n",
    "                            \"Streptococcus mutans\": None,\n",
    "                            \"Streptococcus australis\": None,\n",
    "                            \"Enterobacter asburiae\": None,\n",
    "                            \"Pseudomonas stutzeri\": None,\n",
    "                            \"Kocuria spp, salsicia, rhizophila, kristinae\": None,\n",
    "                            \"Rothia dentocariosa/mucliaginosa\": None,\n",
    "                            \"Pantoea spp\": None,\n",
    "                            \"Pseudomonas citronellolis\": None,\n",
    "                            \"Streptococcus mitis/oralis/cristatus\": None,\n",
    "                            \"Staphylococcus haemolyticus\": None,\n",
    "                            \"Serratia marcescens\": None,\n",
    "                            \"Staphylococcus lugdunensis\": None,\n",
    "                            \"Staphylococcus epidermidis\": None,\n",
    "                            \"Actinomyces neuii\": None,\n",
    "                            \"Streptomyces\": None,\n",
    "                            \"Bacillus licheniformis\": None,\n",
    "                            \"Corynebacterium falsenii\": None,\n",
    "\n",
    "                            # Dysbiotic bacteria — no units\n",
    "                            \"Morganellla morganii\": None,\n",
    "                            \"Providencia stuartii\": None,\n",
    "                            \"Klebsiella pneumonia\": None,\n",
    "                            \"Providencia rettgeri\": None,\n",
    "                            \"Enterobacter cloacae\": None,\n",
    "                            \"Pseudomonas aeruginosa\": None,\n",
    "                            \"Bacillus cereus\": None,\n",
    "                            \"Citrobacter freundii complex\": None,\n",
    "                            \"Arcobacter butzieri\": None,\n",
    "\n",
    "                            # Yeast — no units\n",
    "                            \"Candida parapsilosis\": None,\n",
    "                            \"Rhodotorula glutinis/mucilaginosa\": None,\n",
    "                            \"Candida krusei/inconspicua/lambica\": None,\n",
    "                            \"Saccharomyces cerevisiae/boulardi\": None,\n",
    "                            \"Geotrichum spp.\": None,\n",
    "                            \"Candida pararugosa\": None,\n",
    "                            \"Candida lusitaniae\": None,\n",
    "                            \"Candida zeylanoides\": None,\n",
    "                            \"Trichosporon asahii\": None,\n",
    "                            \"Trichosporon ovoides\": None,\n",
    "                            \"Trichosporon montevideense\": None,\n",
    "                            \"Exophiala dermatitidis\": None,\n",
    "                            \"Cryptococcus spp, not neoformans\": None,\n",
    "                            \"Cryptococcus diffuens\": None,\n",
    "                            \"Candida orthopsilosis\": None,\n",
    "\n",
    "                            # Gut Inflammation\n",
    "                            \"Stool Calprotectin UCSD\": \"µg/g\",\n",
    "                            \"Stool Calprotectin YFH\": None,\n",
    "                            \"Stool Lysozyme\": None,\n",
    "                            \"Stool Lactoferrin\": None,\n",
    "                            \"Stool SIgA\": None,\n",
    "\n",
    "                            # Short Chain Fatty Acids\n",
    "                            \"% SCFA Acetate\": \"%\",\n",
    "                            \"% SCFA Propionate\": \"%\",\n",
    "                            \"% SCFA Valerate\": \"%\",\n",
    "                            \"% SCFA Butyrate\": \"%\",\n",
    "                            \"Total SCFA\": \"mg/mL\",\n",
    "                            \"Butyrate\": \"mg/mL\"\n",
    "\n",
    "}\n",
    "\n",
    "# Mapping for ranges to biomarkers\n",
    "dict_biomarkersToRanges = {\n",
    "                                # Vitals\n",
    "                                \"Weight\": None,\n",
    "\n",
    "                                # LIPID Panel\n",
    "                                \"237\": \"<129\",\n",
    "                                'HDL (High Density Lipoprotein) \"good\"': \"40 to 59\",\n",
    "                                \"VLDL Cholesterol\": \"<30\",\n",
    "                                \"IDL Chloresterol\": \"<20\",\n",
    "                                \"Sum Total Cholesterol\": \"<199\",\n",
    "                                \"Triglycerides\": \"30 to 159\",\n",
    "                                \"UCSD LDL Particle # by NMR\": \"<1135\",\n",
    "                                \"UCSD HDL Particle #\": \">33\",\n",
    "                                \"UCSD Large VLDL Particle #\": \"<2.7\",\n",
    "                                \"UCSD Small LDL Particle #\": \"<634\",\n",
    "                                \"UCSD Large HDL Particle #\": \">4.2\",\n",
    "                                \"UCSD VLDL Particle Size\": \"<46.7\",\n",
    "                                \"UCSD LDL Particle Size\": \">20.7\",\n",
    "                                \"UCSD HDL Particle Size\": \">8.9\",\n",
    "                                \"Real-LDL Size Pattern\": \"A\",\n",
    "                                \"TG/HDL\": \"<2\",\n",
    "                                \"Chol/HDL\": \"<4.5\",\n",
    "                                \"Lp(a) Chloresterol\": \"<10\",\n",
    "                                \"APO A1\": \"94 to 178\",\n",
    "                                \"APOLIPOPROTEIN-(B100)\": \"55 to 140\",\n",
    "                                \"Ratio APO B100/AI\": None,\n",
    "\n",
    "                                # Comprehensive Metabolic Panel – Electrolytes\n",
    "                                \"Sodium\": \"136 to 145\",\n",
    "                                \"Potassium\": \"3.5 to 5.1\",\n",
    "                                \"Chloride\": \"98 to 107\",\n",
    "                                \"Calcium\": \"8.5 to 10.6\",\n",
    "                                \"Phosporus\": \"2.7 to 4.5\",\n",
    "                                \"Magnesium\": \"1.6 to 2.4\",\n",
    "                                \"Zinc\": \"56-134\",\n",
    "                                \"Tot CO2\": \"23-29\",\n",
    "                                \"Bicarbonate\": \"22 to 29\",\n",
    "                                \"Anion Gap\": \"7 to 15\",\n",
    "\n",
    "                                # Comprehensive Metabolic Panel – remainder\n",
    "                                \"Total Protein\": \"6 to 8\",\n",
    "                                \"Albumin\": \"3.5 to 5.2\",\n",
    "                                \"Globulin\": \"2.0-4.5\",\n",
    "                                \"A/G ratio\": \"1.1-2.3\",\n",
    "                                \"GGTP\": \"<60\",\n",
    "                                \"LDH\": \"135 to 225\",\n",
    "                                \"ALT\": \"0 to 41\",\n",
    "                                \"AST\": \"0-40\",\n",
    "                                \"AST/ALT ratio\": None,\n",
    "                                \"Alkaline phosphatase\": \"40 to 120\",\n",
    "                                \"Total Bilirubin\": \"<1.2\",\n",
    "                                \"Direct Bilirubin\": \"<0.2\",\n",
    "\n",
    "                                # Kidney\n",
    "                                \"Blood Urea Nitrogen (BUN)\": \"8 to 23\",\n",
    "                                \"Creatinine\": \"0.67 to 1.17\",\n",
    "                                \"BUN/Creatinine Ratio\": \"10 to 20\",\n",
    "                                \"Uric Acid\": \"4.0-8.5\",\n",
    "\n",
    "                                # Glucose/Insulin Panel\n",
    "                                \"Fasting Glucose\": \"70 to 99\",\n",
    "                                \"Insulin\": \"3 to 19\",\n",
    "                                \"Hemoglobin A1C\": \"4.8 to 5.8\",\n",
    "                                \"HOMA-IR (insulin resistance)\": \"<2.8\",\n",
    "\n",
    "                                # Hormone Panel\n",
    "                                \"Free Testosterone\": \"47-244\",\n",
    "                                \"Testosterone % Free\": None,\n",
    "                                \"Total Testosterone\": \"2.8-8\",\n",
    "                                \"Sex Hormone Binding Globulin (SHBG)\": \"15-100\",\n",
    "                                \"Estradiol\": \"<43\",\n",
    "                                \"DHEA-blood\": \"34 to 249\",\n",
    "                                \"Cortisol Random\": \"6 to 23\",\n",
    "                                \"TSH\": \"0.27 to 4.2\",\n",
    "                                \"Total T3\": \"0.8 to 2.0\",\n",
    "                                \"Thyroxine (T4)\": \"4.5 to 10.9\",\n",
    "\n",
    "                                # Cancer Panel\n",
    "                                \"PSA Screen\": \"<3.99\",\n",
    "                                \"CA-19-9 (Pancreatic Cancer)\": \"30 to 42\",\n",
    "                                \"CEA\": \"<3.8\",\n",
    "\n",
    "                                # Iron Panel\n",
    "                                \"Iron\": \"59 to 158\",\n",
    "                                \"Ferritin\": \"30 to 400\",\n",
    "                                \"Fibrinogen\": \"<460\",\n",
    "                                \"TIBC\": \"148 to 506\",\n",
    "                                \"UIBC\": \"112 to 346\",\n",
    "                                \"Iron saturation\": \"?\",\n",
    "\n",
    "                                # Complete Blood Count (CBC) Panel – Red Blood Cell Details\n",
    "                                \"RBC-\": \"4.6 to 6.1\",\n",
    "                                \"HGB-\": \"13.7 to 17.5\",\n",
    "                                \"HCT\": \"40 to 50\",\n",
    "                                \"PLT-\": \"140 to 370\",\n",
    "                                \"MCV-\": \"79 to 95\",\n",
    "                                \"MCH\": \"26 to 32\",\n",
    "                                \"MCHC-\": \"32 to 36\",\n",
    "                                \"RDW\": \"12 to 14\",\n",
    "                                \"MPV\": \"9.4-12.4\",\n",
    "\n",
    "                                # Complete Blood Count (CBC) Panel  – White Blood Cell Details\n",
    "                                \"WBC-\": \"4 to 10\",\n",
    "                                \"NEU %\": \"41 to 80\",\n",
    "                                \"LYM %\": \"19 to 53\",\n",
    "                                \"EOS %\": \"1 to 7\",\n",
    "                                \"MONO %\": \"5 to 12\",\n",
    "                                \"BASO %\": \"0 to 2\",\n",
    "                                \"Neutrophil Count\": \"1.6 to 7\",\n",
    "                                \"Lymphocyte Count\": \"0.8 to 3.1\",\n",
    "                                \"Monocyte Count\": \"0.2 to 0.8\",\n",
    "                                \"Esoinophil Count\": \"0.1 to 0.5\",\n",
    "                                \"Basophil Count\": \"<0.1\",\n",
    "\n",
    "                                # Inflammation Panel\n",
    "                                \"High Sensitivity CRP (mg/L)\": \"<4.9\",\n",
    "                                \"Lp-PLA2\": \"<200\",\n",
    "                                \"IgG\": \"700-1600\",\n",
    "                                \"IgA\": \"70-400\",\n",
    "                                \"IgM\": \"40-230\",\n",
    "\n",
    "                                # Vitamins, Antioxidants, Amino Acid Panel\n",
    "                                \"Homocysteine\": \"0 to 14\",\n",
    "                                \"CoQ10\": \"0.44-1.64\",\n",
    "                                \"Selenium\": None,\n",
    "                                \"Vitamin E Alpha-Tocopherol\": \"5.5-18\",\n",
    "                                \"Vitamin E Gamma-Tocopherol\": \"0-6\",\n",
    "                                \"Alpha Lipoic Acid\": None,\n",
    "                                \"Vitamin D, 25-Hydroxy\": \"30-80\",\n",
    "                                \"Vitamin B12\": \"232 to 1245\",\n",
    "                                \"Niacin (Vit B3)\": None,\n",
    "\n",
    "                                # Stool Tests\n",
    "                                \"pH\": \"5.8 to 7\",\n",
    "\n",
    "                                # Cultured Bacteria Growth - Expected/Beneficial flora\n",
    "                                \"Bifidobacerium family\": None,\n",
    "                                \"E. coli spp.\": None,\n",
    "                                \"Lactobacillus spp\": None,\n",
    "                                \"Enterococcus spp.\": None,\n",
    "                                \"Bacteroides fraglis group\": None,\n",
    "                                \"Clostridium spp\": None,\n",
    "                                \"Bacteroides family\": None,\n",
    "\n",
    "                                # Cultured Bacteria Growth - Commensal (Imbalanced) flora \n",
    "                                \"Alpha hemolytic strep\": None,\n",
    "                                \"Hemolytic Escherichia coli\": None,\n",
    "                                \"Pseudomonas chlororaphis group\": None,\n",
    "                                \"Gamma Hemolytic Streptococcus\": None,\n",
    "                                \"Hafnia alvei\": None,\n",
    "                                \"Mucoid Escherichia coli\": None,\n",
    "                                \"Beta strep, group B, strep. Agalactiae\": None,\n",
    "                                \"Klebsiella oxytoca\": None,\n",
    "                                \"Bacillus spp, pumilus\": None,\n",
    "                                \"Comamonas testosteroni\": None,\n",
    "                                \"Streptococcus anginosus\": None,\n",
    "                                \"Cellulosimicrobium cellulans/funkeil/marinum\": None,\n",
    "                                \"Lactococcus lactis\": None,\n",
    "                                \"Staphylococcus aureus\": None,\n",
    "                                \"Streptococcus salivarius\": None,\n",
    "                                \"Streptococcus parasanguinis\": None,\n",
    "                                \"Pseudomonas mendocina \": None,\n",
    "                                \"Actinomyces turicensis\": None,\n",
    "                                \"Lysinibacillus spp\": None,\n",
    "                                \"Streptococcus mutans\": None,\n",
    "                                \"Streptococcus australis\": None,\n",
    "                                \"Enterobacter asburiae\": None,\n",
    "                                \"Pseudomonas stutzeri\": None,\n",
    "                                \"Kocuria spp, salsicia, rhizophila, kristinae\": None,\n",
    "                                \"Rothia dentocariosa/mucliaginosa\": None,\n",
    "                                \"Pantoea spp\": None,\n",
    "                                \"Pseudomonas citronellolis\": None,\n",
    "                                \"Streptococcus mitis/oralis/cristatus\": None,\n",
    "                                \"Staphylococcus haemolyticus\": None,\n",
    "                                \"Serratia marcescens\": None,\n",
    "                                \"Staphylococcus lugdunensis\": None,\n",
    "                                \"Staphylococcus epidermidis\": None,\n",
    "                                \"Actinomyces neuii\": None,\n",
    "                                \"Streptomyces\": None,\n",
    "                                \"Bacillus licheniformis\": None,\n",
    "                                \"Corynebacterium falsenii\": None,\n",
    "\n",
    "                                # Dysbiotic bacteria — no ranges\n",
    "                                \"Morganellla morganii\": None,\n",
    "                                \"Providencia stuartii\": None,\n",
    "                                \"Klebsiella pneumonia\": None,\n",
    "                                \"Providencia rettgeri\": None,\n",
    "                                \"Enterobacter cloacae\": None,\n",
    "                                \"Pseudomonas aeruginosa\": None,\n",
    "                                \"Bacillus cereus\": None,\n",
    "                                \"Citrobacter freundii complex\": None,\n",
    "                                \"Arcobacter butzieri\": None,\n",
    "\n",
    "                                # Yeast — no ranges\n",
    "                                \"Candida parapsilosis\": None,\n",
    "                                \"Rhodotorula glutinis/mucilaginosa\": None,\n",
    "                                \"Candida krusei/inconspicua/lambica\": None,\n",
    "                                \"Saccharomyces cerevisiae/boulardi\": None,\n",
    "                                \"Geotrichum spp.\": None,\n",
    "                                \"Candida pararugosa\": None,\n",
    "                                \"Candida lusitaniae\": None,\n",
    "                                \"Candida zeylanoides\": None,\n",
    "                                \"Trichosporon asahii\": None,\n",
    "                                \"Trichosporon ovoides\": None,\n",
    "                                \"Trichosporon montevideense\": None,\n",
    "                                \"Exophiala dermatitidis\": None,\n",
    "                                \"Cryptococcus spp, not neoformans\": None,\n",
    "                                \"Cryptococcus diffuens\": None,\n",
    "                                \"Candida orthopsilosis\": None,\n",
    "\n",
    "                                # Gut Inflammation\n",
    "                                \"Stool Calprotectin UCSD\": \"<50\",\n",
    "                                \"Stool Calprotectin YFH\": \"<80\",\n",
    "                                \"Stool Lysozyme\": \"<500\",\n",
    "                                \"Stool Lactoferrin\": \"<7.3\",\n",
    "                                \"Stool SIgA\": \"30-275\",\n",
    "\n",
    "                                # Short Chain Fatty Acids\n",
    "                                \"% SCFA Acetate\": \"50 to 72\",\n",
    "                                \"% SCFA Propionate\": \"11 to 25\",\n",
    "                                \"% SCFA Valerate\": \"1 to 5\",\n",
    "                                \"% SCFA Butyrate\": \"11 to 32\",\n",
    "                                \"Total SCFA\": \"5 to 16\",\n",
    "                                \"Butyrate\": \"0.8-4\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "f901aa3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop Columns with Low, High, Average,\tNumber\t\n",
    "df_longBiomarkers= df_biomarkers.drop(columns=['Low', 'High', 'Average', 'Number'])\n",
    "\n",
    "# Collect columns\n",
    "cols = list(df_longBiomarkers.columns)\n",
    "\n",
    "# Create columns for biomarker, dates, facility\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "d3033054",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_biometricPanel = find_exact(\"Date collected\") or cols[0]  # biomarker/test label\n",
    "col_units = find_exact(\"Units\")\n",
    "col_range = find_exact(\"Good Range\")\n",
    "cols_meta = {c for c in [col_biometricPanel, col_units, col_range] if c is not None}\n",
    "\n",
    "cols_drop = {\"low\", \"high\", \"average\", \"number\"}\n",
    "\n",
    "# Collect date column\n",
    "rgx_date = re.compile(r\"\\s*\\d{1,2}/\\d{1,2}/\\d{2,4}\\s*\")  # must look like M/D/YYYY\n",
    "col_date = [\n",
    "                c for c in df_biomarkers.columns\n",
    "                if (c not in cols_meta)\n",
    "                and (str(c).strip().lower() not in cols_drop)\n",
    "                and rgx_date.fullmatch(str(c))  # must look like M/D/YYYY\n",
    "]\n",
    "\n",
    "col_date = [c for c in col_date if not str(c).lower().startswith(\"unnamed\")]\n",
    "\n",
    "# Facility column\n",
    "row_one = df_biomarkers.iloc[0]\n",
    "dict_facilityByDate = {\n",
    "                        dc: (None if is_blank(row_one.get(dc)) \n",
    "                        else str(row_one.get(dc)).strip())\n",
    "                        for dc in col_date\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "2c52a8e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['Weight',\n",
       "  'Blood Tests',\n",
       "  'LIPID Panel',\n",
       "  '237',\n",
       "  'HDL (High Density Lipoprotein) \"good\"',\n",
       "  'VLDL Cholesterol',\n",
       "  'IDL Chloresterol',\n",
       "  'Sum Total Cholesterol',\n",
       "  'Triglycerides',\n",
       "  'UCSD LDL Particle # by NMR',\n",
       "  'UCSD HDL Particle #',\n",
       "  'UCSD Large VLDL Particle #',\n",
       "  'UCSD Small LDL Particle #',\n",
       "  'UCSD Large HDL Particle #',\n",
       "  'UCSD VLDL Particle Size',\n",
       "  'UCSD LDL Particle Size',\n",
       "  'UCSD HDL Particle Size',\n",
       "  'Real-LDL Size Pattern',\n",
       "  'TG/HDL',\n",
       "  'Chol/HDL',\n",
       "  'Lp(a) Chloresterol',\n",
       "  'APO A1',\n",
       "  'APOLIPOPROTEIN-(B100)',\n",
       "  'Ratio APO B100/AI',\n",
       "  'Comprehensive Metabolic Panel',\n",
       "  'Electolytes',\n",
       "  'Sodium',\n",
       "  'Potassium',\n",
       "  'Chloride',\n",
       "  'Calcium',\n",
       "  'Phosporus',\n",
       "  'Magnesium',\n",
       "  'Zinc',\n",
       "  'Tot CO2',\n",
       "  'Bicarbonate',\n",
       "  'Anion Gap',\n",
       "  'Liver',\n",
       "  'Total Protein',\n",
       "  'Albumin',\n",
       "  'Globulin',\n",
       "  'A/G ratio',\n",
       "  'GGTP',\n",
       "  'LDH',\n",
       "  'ALT',\n",
       "  'AST',\n",
       "  'AST/ALT ratio',\n",
       "  'Alkaline phosphatase',\n",
       "  'Total Bilirubin',\n",
       "  'Direct Bilirubin',\n",
       "  'Kidney',\n",
       "  'Blood Urea Nitrogen (BUN)',\n",
       "  'Creatinine',\n",
       "  'BUN/Creatinine Ratio',\n",
       "  'Uric Acid',\n",
       "  'Glucose/Insulin Panel',\n",
       "  'Fasting Glucose',\n",
       "  'Insulin',\n",
       "  'Hemoglobin A1C',\n",
       "  'HOMA-IR (insulin resistance)',\n",
       "  'Hormone Panel',\n",
       "  'Free Testosterone',\n",
       "  'Testosterone % Free',\n",
       "  'Total Testosterone',\n",
       "  'Sex Hormone Binding Globulin (SHBG)',\n",
       "  'Estradiol',\n",
       "  'DHEA-blood',\n",
       "  'Cortisol Random',\n",
       "  'TSH',\n",
       "  'Total T3',\n",
       "  'Thyroxine (T4)',\n",
       "  'Cancer Panel',\n",
       "  'PSA Screen',\n",
       "  'CA-19-9 (Pancreatic Cancer)',\n",
       "  'CEA',\n",
       "  'Iron Panel',\n",
       "  'Iron',\n",
       "  'Ferritin',\n",
       "  'Fibrinogen',\n",
       "  'TIBC',\n",
       "  'UIBC',\n",
       "  'Iron saturation',\n",
       "  'Complete Blood Count (CBC) Panel',\n",
       "  'Red Blood Cell Details',\n",
       "  'RBC-',\n",
       "  'HGB-',\n",
       "  'HCT',\n",
       "  'PLT-',\n",
       "  'MCV-',\n",
       "  'MCH',\n",
       "  'MCHC-',\n",
       "  'RDW',\n",
       "  'MPV',\n",
       "  'White Blood Cell Details',\n",
       "  'WBC-',\n",
       "  'NEU %',\n",
       "  'LYM %',\n",
       "  'EOS %',\n",
       "  'MONO %',\n",
       "  'BASO %',\n",
       "  'Neutrophil Count',\n",
       "  'Lymphocyte Count',\n",
       "  'Monocyte Count',\n",
       "  'Esoinophil Count',\n",
       "  'Basophil Count',\n",
       "  'Inflammation Panel',\n",
       "  'High Sensitivity CRP (mg/L)',\n",
       "  'Lp-PLA2',\n",
       "  'IgG',\n",
       "  'IgA',\n",
       "  'IgM',\n",
       "  'Vitamins, Antioxidants, Amino Acid Panel',\n",
       "  'Homocysteine',\n",
       "  'CoQ10',\n",
       "  'Selenium',\n",
       "  'Vitamin E Alpha-Tocopherol',\n",
       "  'Vitamin E Gamma-Tocopherol',\n",
       "  'Alpha Lipoic Acid',\n",
       "  'Vitamin D, 25-Hydroxy',\n",
       "  'Vitamin B12',\n",
       "  'Niacin (Vit B3)',\n",
       "  'Stool Tests',\n",
       "  'pH',\n",
       "  'Cultured Bacteria Growth',\n",
       "  'Expected/Beneficial flora',\n",
       "  'Bifidobacerium family',\n",
       "  'E. coli spp.',\n",
       "  'Lactobacillus spp',\n",
       "  'Enterococcus spp.',\n",
       "  'Bacteroides fraglis group',\n",
       "  'Clostridium spp',\n",
       "  'Bacteroides family',\n",
       "  'Commensal (Imbalanced) flora',\n",
       "  'Alpha hemolytic strep',\n",
       "  'Hemolytic Escherichia coli',\n",
       "  'Pseudomonas chlororaphis group',\n",
       "  'Gamma Hemolytic Streptococcus',\n",
       "  'Hafnia alvei',\n",
       "  'Mucoid Escherichia coli',\n",
       "  'Beta strep, group B, strep. Agalactiae',\n",
       "  'Klebsiella oxytoca',\n",
       "  'Bacillus spp, pumilus',\n",
       "  'Comamonas testosteroni',\n",
       "  'Streptococcus anginosus',\n",
       "  'Cellulosimicrobium cellulans/funkeil/marinum',\n",
       "  'Lactococcus lactis',\n",
       "  'Staphylococcus aureus',\n",
       "  'Streptococcus salivarius',\n",
       "  'Streptococcus parasanguinis',\n",
       "  'Pseudomonas mendocina',\n",
       "  'Actinomyces turicensis',\n",
       "  'Lysinibacillus spp',\n",
       "  'Streptococcus mutans',\n",
       "  'Streptococcus australis',\n",
       "  'Enterobacter asburiae',\n",
       "  'Pseudomonas stutzeri',\n",
       "  'Kocuria spp, salsicia, rhizophila, kristinae',\n",
       "  'Rothia dentocariosa/mucliaginosa',\n",
       "  'Pantoea spp',\n",
       "  'Pseudomonas citronellolis',\n",
       "  'Streptococcus mitis/oralis/cristatus',\n",
       "  'Staphylococcus haemolyticus',\n",
       "  'Serratia marcescens',\n",
       "  'Staphylococcus lugdunensis',\n",
       "  'Staphylococcus epidermidis',\n",
       "  'Actinomyces neuii',\n",
       "  'Streptomyces',\n",
       "  'Bacillus licheniformis',\n",
       "  'Corynebacterium falsenii',\n",
       "  'Dysbiotic bacteria',\n",
       "  'Morganellla morganii',\n",
       "  'Providencia stuartii',\n",
       "  'Klebsiella pneumonia',\n",
       "  'Providencia rettgeri',\n",
       "  'Enterobacter cloacae',\n",
       "  'Pseudomonas aeruginosa',\n",
       "  'Bacillus cereus',\n",
       "  'Citrobacter freundii complex',\n",
       "  'Arcobacter butzieri',\n",
       "  'Yeast',\n",
       "  'Candida parapsilosis',\n",
       "  'Rhodotorula glutinis/mucilaginosa',\n",
       "  'Candida krusei/inconspicua/lambica',\n",
       "  'Saccharomyces cerevisiae/boulardi',\n",
       "  'Geotrichum spp.',\n",
       "  'Candida pararugosa',\n",
       "  'Candida lusitaniae',\n",
       "  'Candida zeylanoides',\n",
       "  'Trichosporon asahii',\n",
       "  'Trichosporon ovoides',\n",
       "  'Trichosporon montevideense',\n",
       "  'Exophiala dermatitidis',\n",
       "  'Cryptococcus spp, not neoformans',\n",
       "  'Cryptococcus diffuens',\n",
       "  'Candida orthopsilosis',\n",
       "  'Gut Inflammation',\n",
       "  'Stool Calprotectin UCSD',\n",
       "  'Stool Calprotectin YFH',\n",
       "  'Stool Lysozyme',\n",
       "  'Stool Lactoferrin',\n",
       "  'Stool SIgA',\n",
       "  'Short Chain Fatty Acids',\n",
       "  '% SCFA Acetate',\n",
       "  '% SCFA Propionate',\n",
       "  '% SCFA Valerate',\n",
       "  '% SCFA Butyrate',\n",
       "  'Total SCFA',\n",
       "  'Butyrate'],\n",
       " ['LIPID Panel',\n",
       "  'Comprehensive Metabolic Panel',\n",
       "  'Kidney',\n",
       "  'Glucose/Insulin Panel',\n",
       "  'Hormone Panel',\n",
       "  'Cancer Panel',\n",
       "  'Iron Panel',\n",
       "  'Inflammation Panel',\n",
       "  'Vitamins, Antioxidants, Amino Acid Panel',\n",
       "  'Expected/Beneficial flora',\n",
       "  'Commensal (Imbalanced) flora',\n",
       "  'Dysbiotic bacteria',\n",
       "  'Yeast',\n",
       "  'Gut Inflammation',\n",
       "  'Short Chain Fatty Acids'])"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Collect Biomarkers and Panel Columns from the Date Collected column\n",
    "\n",
    "lst_panels = [\n",
    "    \"Vitals\",\n",
    "    \"LIPID Panel\",\n",
    "    \"Comprehensive Metabolic Panel\",\n",
    "    \"Kidney\",\n",
    "    \"Glucose/Insulin Panel\",\n",
    "    \"Hormone Panel\",\n",
    "    \"Cancer Panel\",\n",
    "    \"Iron Panel\",\n",
    "    \"CBC: Red Blood Cell Details\",\n",
    "    \"CBC: White Blood Cell Details\",\n",
    "    \"Inflammation Panel\",\n",
    "    \"Vitamins, Antioxidants, Amino Acid Panel\",\n",
    "    \"Stool tests\",\n",
    "    \"Expected/Beneficial flora\",\n",
    "    \"Commensal (Imbalanced) flora\",\n",
    "    \"Dysbiotic bacteria\",\n",
    "    \"Yeast\",\n",
    "    \"Gut Inflammation\",\n",
    "    \"Short Chain Fatty Acids\",\n",
    "]\n",
    "\n",
    "# --- pull biomarkers from the \"Date collected\" column and make clean lists ---\n",
    "\n",
    "# 0) left column cleaned\n",
    "labels = [clean_label(x) for x in df_biomarkers[col_biometricPanel].tolist()]\n",
    "\n",
    "# 1) drop blanks/separators\n",
    "labels_noblank = [x for x in labels if x is not None]\n",
    "\n",
    "# 2) panels found (in-order, unique) limited to your fixed lst_panels\n",
    "lst_panels_found = uniq_order([x for x in labels_noblank if x in lst_panels])\n",
    "\n",
    "labels_noblank, lst_panels_found \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "2df43535",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/nf/dp5pbtr55rn8mww76lsrkt800000gn/T/ipykernel_80014/766801454.py:22: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead\n",
      "  df_longBiomarkers[\"Value\"] = pd.to_numeric(df_longBiomarkers[\"Value\"], errors=\"ignore\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Biomarker</th>\n",
       "      <th>Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1993-05-01</td>\n",
       "      <td>Sum Total Cholesterol</td>\n",
       "      <td>152.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1996-11-27</td>\n",
       "      <td>A/G ratio</td>\n",
       "      <td>1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1996-11-27</td>\n",
       "      <td>AST</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1996-11-27</td>\n",
       "      <td>Albumin</td>\n",
       "      <td>4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1996-11-27</td>\n",
       "      <td>Alkaline phosphatase</td>\n",
       "      <td>64.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22857</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>Vitamin D, 25-Hydroxy</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22858</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>Vitamin E Alpha-Tocopherol</td>\n",
       "      <td>7.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22859</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>Vitamin E Gamma-Tocopherol</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22860</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>WBC-</td>\n",
       "      <td>4.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22861</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>Weight</td>\n",
       "      <td>180.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22862 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Date                   Biomarker  Value\n",
       "0     1993-05-01       Sum Total Cholesterol  152.0\n",
       "1     1996-11-27                   A/G ratio    1.3\n",
       "2     1996-11-27                         AST   37.0\n",
       "3     1996-11-27                     Albumin    4.2\n",
       "4     1996-11-27        Alkaline phosphatase   64.0\n",
       "...          ...                         ...    ...\n",
       "22857 2025-06-13       Vitamin D, 25-Hydroxy     69\n",
       "22858 2025-06-13  Vitamin E Alpha-Tocopherol    7.9\n",
       "22859 2025-06-13  Vitamin E Gamma-Tocopherol    0.6\n",
       "22860 2025-06-13                        WBC-    4.4\n",
       "22861 2025-06-13                      Weight  180.8\n",
       "\n",
       "[22862 rows x 3 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract Biomarkrker value which are the column/row, date, biomarker\n",
    "\n",
    "# rows that are biomarkers (not panel headers)\n",
    "idx_biomarker_rows = [\n",
    "                        i for i, x in enumerate(labels)\n",
    "                        if (x is not None) and (x in dict_biomarkerToPanels)\n",
    "]\n",
    "\n",
    "# build long rows directly (no melt, no wide step)\n",
    "records = []\n",
    "for i in idx_biomarker_rows:\n",
    "    biomarker = labels[i]  # already cleaned\n",
    "    row = df_biomarkers.loc[i, col_date]\n",
    "    for dc, val in row.items():\n",
    "        if not is_blank(val):\n",
    "            records.append({\"Date\": dc, \"Biomarker\": biomarker, \"Value\": val})\n",
    "\n",
    "df_longBiomarkers = pd.DataFrame.from_records(records)\n",
    "\n",
    "# tidy types and order (still long)\n",
    "df_longBiomarkers[\"Date\"]  = pd.to_datetime(df_longBiomarkers[\"Date\"], errors=\"coerce\")\n",
    "df_longBiomarkers[\"Value\"] = pd.to_numeric(df_longBiomarkers[\"Value\"], errors=\"ignore\")\n",
    "df_longBiomarkers = df_longBiomarkers.sort_values([\"Date\",\"Biomarker\"]).reset_index(drop=True)\n",
    "df_longBiomarkers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "b37ca2db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Biomarker</th>\n",
       "      <th>Value</th>\n",
       "      <th>Units</th>\n",
       "      <th>Panel</th>\n",
       "      <th>GoodRange</th>\n",
       "      <th>Facility</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1993-05-01</td>\n",
       "      <td>Sum Total Cholesterol</td>\n",
       "      <td>152.0</td>\n",
       "      <td>mg/dL</td>\n",
       "      <td>LIPID Panel</td>\n",
       "      <td>&lt;199</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1996-11-27</td>\n",
       "      <td>A/G ratio</td>\n",
       "      <td>1.3</td>\n",
       "      <td>None</td>\n",
       "      <td>Comprehensive Metabolic Panel</td>\n",
       "      <td>1.1-2.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1996-11-27</td>\n",
       "      <td>AST</td>\n",
       "      <td>37.0</td>\n",
       "      <td>U/L</td>\n",
       "      <td>Comprehensive Metabolic Panel</td>\n",
       "      <td>0-40</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1996-11-27</td>\n",
       "      <td>Albumin</td>\n",
       "      <td>4.2</td>\n",
       "      <td>g/dL</td>\n",
       "      <td>Comprehensive Metabolic Panel</td>\n",
       "      <td>3.5 to 5.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1996-11-27</td>\n",
       "      <td>Alkaline phosphatase</td>\n",
       "      <td>64.0</td>\n",
       "      <td>U/L</td>\n",
       "      <td>Comprehensive Metabolic Panel</td>\n",
       "      <td>40 to 120</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22857</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>Vitamin D, 25-Hydroxy</td>\n",
       "      <td>69</td>\n",
       "      <td>ng/mL</td>\n",
       "      <td>Vitamins, Antioxidants, Amino Acid Panel</td>\n",
       "      <td>30-80</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22858</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>Vitamin E Alpha-Tocopherol</td>\n",
       "      <td>7.9</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Vitamins, Antioxidants, Amino Acid Panel</td>\n",
       "      <td>5.5-18</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22859</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>Vitamin E Gamma-Tocopherol</td>\n",
       "      <td>0.6</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Vitamins, Antioxidants, Amino Acid Panel</td>\n",
       "      <td>0-6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22860</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>WBC-</td>\n",
       "      <td>4.4</td>\n",
       "      <td>10**3/mL</td>\n",
       "      <td>CBC: White Blood Cell Details</td>\n",
       "      <td>4 to 10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22861</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>Weight</td>\n",
       "      <td>180.8</td>\n",
       "      <td>lbs</td>\n",
       "      <td>Vitals</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22862 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Date                   Biomarker  Value     Units  \\\n",
       "0     1993-05-01       Sum Total Cholesterol  152.0     mg/dL   \n",
       "1     1996-11-27                   A/G ratio    1.3      None   \n",
       "2     1996-11-27                         AST   37.0       U/L   \n",
       "3     1996-11-27                     Albumin    4.2      g/dL   \n",
       "4     1996-11-27        Alkaline phosphatase   64.0       U/L   \n",
       "...          ...                         ...    ...       ...   \n",
       "22857 2025-06-13       Vitamin D, 25-Hydroxy     69     ng/mL   \n",
       "22858 2025-06-13  Vitamin E Alpha-Tocopherol    7.9      mg/L   \n",
       "22859 2025-06-13  Vitamin E Gamma-Tocopherol    0.6      mg/L   \n",
       "22860 2025-06-13                        WBC-    4.4  10**3/mL   \n",
       "22861 2025-06-13                      Weight  180.8       lbs   \n",
       "\n",
       "                                          Panel   GoodRange Facility  \n",
       "0                                   LIPID Panel        <199      NaN  \n",
       "1                 Comprehensive Metabolic Panel     1.1-2.3      NaN  \n",
       "2                 Comprehensive Metabolic Panel        0-40      NaN  \n",
       "3                 Comprehensive Metabolic Panel  3.5 to 5.2      NaN  \n",
       "4                 Comprehensive Metabolic Panel   40 to 120      NaN  \n",
       "...                                         ...         ...      ...  \n",
       "22857  Vitamins, Antioxidants, Amino Acid Panel       30-80      NaN  \n",
       "22858  Vitamins, Antioxidants, Amino Acid Panel      5.5-18      NaN  \n",
       "22859  Vitamins, Antioxidants, Amino Acid Panel         0-6      NaN  \n",
       "22860             CBC: White Blood Cell Details     4 to 10      NaN  \n",
       "22861                                    Vitals        None      NaN  \n",
       "\n",
       "[22862 rows x 7 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make units, panel, ranges, and facilitie columns by mapping\n",
    "\n",
    "df_longBiomarkers[\"Units\"] = df_longBiomarkers[\"Biomarker\"].map(dict_unitsToBiomarkers)\n",
    "df_longBiomarkers[\"Panel\"] = df_longBiomarkers[\"Biomarker\"].map(dict_biomarkerToPanels)\n",
    "df_longBiomarkers[\"GoodRange\"] = df_longBiomarkers[\"Biomarker\"].map(dict_biomarkersToRanges)\n",
    "df_longBiomarkers[\"Facility\"] = df_longBiomarkers[\"Date\"].map(dict_facilityByDate)\n",
    "df_longBiomarkers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "943e86d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Facility\n",
      "NaN    22862\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# 1) rebuild facility dict with normalized datetime keys; skip bad/blank keys\n",
    "_fixed = {}\n",
    "for k, v in dict_facilityByDate.items():\n",
    "    if is_blank(k):\n",
    "        continue\n",
    "    ts = pd.to_datetime(str(k).strip(), errors=\"coerce\")\n",
    "    if pd.isna(ts):\n",
    "        continue\n",
    "    _fixed[ts.normalize()] = (None if is_blank(v) else str(v).strip())\n",
    "dict_facilityByDate = _fixed  # overwrite with safe version\n",
    "\n",
    "# 2) map facility WITHOUT changing df_longBiomarkers[\"Date\"]\n",
    "df_biomarkers[\"Facility\"] = pd.to_datetime(\n",
    "    df_longBiomarkers[\"Date\"], errors=\"coerce\"\n",
    ").dt.normalize().map(dict_facilityByDate)\n",
    "\n",
    "# quick check\n",
    "print(df_longBiomarkers[\"Facility\"].value_counts(dropna=False).head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "07d33098",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: count, dtype: int64)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_longBiomarkers[\"Facility\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "b553c917",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved Data/LLM_Biomarkers_Long.csv\n"
     ]
    }
   ],
   "source": [
    "df_longBiomarkers.to_csv(fp_biomarkersLong,index=False)\n",
    "print(f\"Saved {fp_biomarkersLong}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "587a83ec",
   "metadata": {},
   "source": [
    "## Make Wide Format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "31e0b9e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Biomarker_Unit</th>\n",
       "      <th>Date</th>\n",
       "      <th>% SCFA Acetate</th>\n",
       "      <th>% SCFA Butyrate</th>\n",
       "      <th>% SCFA Propionate</th>\n",
       "      <th>% SCFA Valerate</th>\n",
       "      <th>237_mg/dL</th>\n",
       "      <th>A/G ratio_None</th>\n",
       "      <th>ALT_U/L</th>\n",
       "      <th>APO A1_mg/dL</th>\n",
       "      <th>APOLIPOPROTEIN-(B100)_mg/dL</th>\n",
       "      <th>...</th>\n",
       "      <th>Uric Acid_None</th>\n",
       "      <th>VLDL Cholesterol_mg/dL</th>\n",
       "      <th>Vitamin B12_pg/mL</th>\n",
       "      <th>Vitamin D, 25-Hydroxy_ng/mL</th>\n",
       "      <th>Vitamin E Alpha-Tocopherol_mg/L</th>\n",
       "      <th>Vitamin E Gamma-Tocopherol_mg/L</th>\n",
       "      <th>WBC-_10**3/mL</th>\n",
       "      <th>Weight_lbs</th>\n",
       "      <th>Zinc_µg/dL</th>\n",
       "      <th>pH_None</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1993-05-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1996-11-27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1997-05-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>6.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1998-04-15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1999-04-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>2025-02-05</td>\n",
       "      <td>47</td>\n",
       "      <td>24</td>\n",
       "      <td>22</td>\n",
       "      <td>7.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.7</td>\n",
       "      <td>178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>2025-02-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>2025-02-28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.5</td>\n",
       "      <td>179.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>2025-05-20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>182.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>2025-06-13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1743</td>\n",
       "      <td>69</td>\n",
       "      <td>7.9</td>\n",
       "      <td>0.6</td>\n",
       "      <td>4.4</td>\n",
       "      <td>180.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>332 rows × 184 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Biomarker_Unit       Date % SCFA Acetate % SCFA Butyrate % SCFA Propionate  \\\n",
       "0              1993-05-01            NaN             NaN               NaN   \n",
       "1              1996-11-27            NaN             NaN               NaN   \n",
       "2              1997-05-08            NaN             NaN               NaN   \n",
       "3              1998-04-15            NaN             NaN               NaN   \n",
       "4              1999-04-01            NaN             NaN               NaN   \n",
       "..                    ...            ...             ...               ...   \n",
       "327            2025-02-05             47              24                22   \n",
       "328            2025-02-12            NaN             NaN               NaN   \n",
       "329            2025-02-28            NaN             NaN               NaN   \n",
       "330            2025-05-20            NaN             NaN               NaN   \n",
       "331            2025-06-13            NaN             NaN               NaN   \n",
       "\n",
       "Biomarker_Unit % SCFA Valerate 237_mg/dL A/G ratio_None ALT_U/L APO A1_mg/dL  \\\n",
       "0                          NaN       NaN            NaN     NaN          NaN   \n",
       "1                          NaN       NaN            1.3     NaN          NaN   \n",
       "2                          NaN       NaN            NaN     NaN          NaN   \n",
       "3                          NaN       NaN            1.4     NaN          NaN   \n",
       "4                          NaN       NaN            NaN     NaN          NaN   \n",
       "..                         ...       ...            ...     ...          ...   \n",
       "327                        7.2       NaN            NaN      22          NaN   \n",
       "328                        NaN       NaN            NaN      14          NaN   \n",
       "329                        NaN       NaN            NaN     NaN          NaN   \n",
       "330                        NaN       NaN            NaN      17          NaN   \n",
       "331                        NaN        57            NaN      15          NaN   \n",
       "\n",
       "Biomarker_Unit APOLIPOPROTEIN-(B100)_mg/dL  ... Uric Acid_None  \\\n",
       "0                                      NaN  ...            NaN   \n",
       "1                                      NaN  ...            6.0   \n",
       "2                                      NaN  ...            6.3   \n",
       "3                                      NaN  ...            NaN   \n",
       "4                                      NaN  ...            NaN   \n",
       "..                                     ...  ...            ...   \n",
       "327                                    NaN  ...            NaN   \n",
       "328                                    NaN  ...            NaN   \n",
       "329                                    NaN  ...            NaN   \n",
       "330                                    NaN  ...            NaN   \n",
       "331                                    NaN  ...            NaN   \n",
       "\n",
       "Biomarker_Unit VLDL Cholesterol_mg/dL Vitamin B12_pg/mL  \\\n",
       "0                                 NaN               NaN   \n",
       "1                                 NaN               NaN   \n",
       "2                                 NaN               NaN   \n",
       "3                                 NaN               NaN   \n",
       "4                                 NaN               NaN   \n",
       "..                                ...               ...   \n",
       "327                               NaN               NaN   \n",
       "328                               NaN               NaN   \n",
       "329                               NaN               NaN   \n",
       "330                               NaN               NaN   \n",
       "331                               NaN              1743   \n",
       "\n",
       "Biomarker_Unit Vitamin D, 25-Hydroxy_ng/mL Vitamin E Alpha-Tocopherol_mg/L  \\\n",
       "0                                      NaN                             NaN   \n",
       "1                                      NaN                             NaN   \n",
       "2                                      NaN                             NaN   \n",
       "3                                      NaN                             NaN   \n",
       "4                                      NaN                             NaN   \n",
       "..                                     ...                             ...   \n",
       "327                                    NaN                             NaN   \n",
       "328                                     49                             NaN   \n",
       "329                                    NaN                             NaN   \n",
       "330                                     64                             NaN   \n",
       "331                                     69                             7.9   \n",
       "\n",
       "Biomarker_Unit Vitamin E Gamma-Tocopherol_mg/L WBC-_10**3/mL Weight_lbs  \\\n",
       "0                                          NaN           NaN        NaN   \n",
       "1                                          NaN           8.7        NaN   \n",
       "2                                          NaN           6.3        NaN   \n",
       "3                                          NaN           6.4        NaN   \n",
       "4                                          NaN           8.7        NaN   \n",
       "..                                         ...           ...        ...   \n",
       "327                                        NaN           6.7        178   \n",
       "328                                        NaN             4        179   \n",
       "329                                        NaN           5.5      179.6   \n",
       "330                                        NaN             4      182.5   \n",
       "331                                        0.6           4.4      180.8   \n",
       "\n",
       "Biomarker_Unit Zinc_µg/dL pH_None  \n",
       "0                     NaN     NaN  \n",
       "1                     NaN     NaN  \n",
       "2                     NaN     NaN  \n",
       "3                     NaN     NaN  \n",
       "4                     NaN     NaN  \n",
       "..                    ...     ...  \n",
       "327                   NaN     6.5  \n",
       "328                   NaN     NaN  \n",
       "329                   NaN     NaN  \n",
       "330                   NaN     NaN  \n",
       "331                   NaN     NaN  \n",
       "\n",
       "[332 rows x 184 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creat wide dataframe from longBioDairyLS\n",
    "\n",
    "# Change column labels to biomarker_unit excpet if \"%\" contained in biomarker\n",
    "df_longBiomarkers[\"Biomarker_Unit\"] = df_longBiomarkers.apply(\n",
    "                                                                lambda x: f\"{x['Biomarker']}_{x['Units']}\"\n",
    "                                                                if \"%\" not in x[\"Biomarker\"] else x[\"Biomarker\"],\n",
    "                                                                axis=1\n",
    ")\n",
    "\n",
    "# Pivot to make date the rows and biomarkers the columns\n",
    "df_wideBiomarkers = df_longBiomarkers.pivot(\n",
    "                                            index=\"Date\", \n",
    "                                            columns=\"Biomarker_Unit\", \n",
    "                                            values=\"Value\"\n",
    ").reset_index()\n",
    "df_wideBiomarkers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b2193200",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved Data/LLM_Biomarkers_Wide.csv\n"
     ]
    }
   ],
   "source": [
    "df_wideBiomarkers.to_csv(fp_biomarkersWide, index=False)\n",
    "print(f\"Saved {fp_biomarkersWide}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2237f18e",
   "metadata": {},
   "source": [
    "# Gut Microbiome Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "1a2891c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LS Gut Microbiome</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Unnamed: 3</th>\n",
       "      <th>Unnamed: 4</th>\n",
       "      <th>Unnamed: 5</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 230</th>\n",
       "      <th>Unnamed: 231</th>\n",
       "      <th>Unnamed: 232</th>\n",
       "      <th>Unnamed: 233</th>\n",
       "      <th>Unnamed: 234</th>\n",
       "      <th>Unnamed: 235</th>\n",
       "      <th>Unnamed: 236</th>\n",
       "      <th>Unnamed: 237</th>\n",
       "      <th>Unnamed: 238</th>\n",
       "      <th>Unnamed: 239</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sequenced in Knight Lab from LS Frozen Stool S...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Date stool sample collected</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Max</td>\n",
       "      <td>Average</td>\n",
       "      <td>12/28/11</td>\n",
       "      <td>4/3/12</td>\n",
       "      <td>8/7/12</td>\n",
       "      <td>11/6/12</td>\n",
       "      <td>1/26/13</td>\n",
       "      <td>2/8/13</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16S Relative Abundance</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sum of 16S Relative Abundances</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.15</td>\n",
       "      <td>97.89</td>\n",
       "      <td>97.98</td>\n",
       "      <td>98.39</td>\n",
       "      <td>97.89</td>\n",
       "      <td>98.15</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Dominate species from Metagenomics comparison</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>Escherichia coli Max HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.6032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>Morganella morganii Ave HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000108823529411765</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>Morganella morganii Max HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>Bifidobacterium longum Ave HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.108320588235294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>Bifidobacterium longum Max HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.2247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>286 rows × 240 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    LS Gut Microbiome   \\\n",
       "0    Sequenced in Knight Lab from LS Frozen Stool S...   \n",
       "1                          Date stool sample collected   \n",
       "2                               16S Relative Abundance   \n",
       "3                       Sum of 16S Relative Abundances   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "281                            Escherichia coli Max HE   \n",
       "282                         Morganella morganii Ave HE   \n",
       "283                         Morganella morganii Max HE   \n",
       "284                      Bifidobacterium longum Ave HE   \n",
       "285                      Bifidobacterium longum Max HE   \n",
       "\n",
       "                                        Unnamed: 1 Unnamed: 2 Unnamed: 3  \\\n",
       "0                                              NaN        NaN        NaN   \n",
       "1                                              NaN        Max    Average   \n",
       "2                                              NaN        NaN        NaN   \n",
       "3                                              NaN        NaN        NaN   \n",
       "4    Dominate species from Metagenomics comparison        NaN        NaN   \n",
       "..                                             ...        ...        ...   \n",
       "281                                            NaN        NaN        NaN   \n",
       "282                                            NaN        NaN        NaN   \n",
       "283                                            NaN        NaN        NaN   \n",
       "284                                            NaN        NaN        NaN   \n",
       "285                                            NaN        NaN        NaN   \n",
       "\n",
       "               Unnamed: 4 Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8  \\\n",
       "0                     NaN        NaN        NaN        NaN        NaN   \n",
       "1                12/28/11     4/3/12     8/7/12    11/6/12    1/26/13   \n",
       "2                     16S        16S        16S        16S        16S   \n",
       "3                   97.15      97.89      97.98      98.39      97.89   \n",
       "4                     NaN        NaN        NaN        NaN        NaN   \n",
       "..                    ...        ...        ...        ...        ...   \n",
       "281                0.6032        NaN        NaN        NaN        NaN   \n",
       "282  0.000108823529411765        NaN        NaN        NaN        NaN   \n",
       "283                0.0004        NaN        NaN        NaN        NaN   \n",
       "284     0.108320588235294        NaN        NaN        NaN        NaN   \n",
       "285                1.2247        NaN        NaN        NaN        NaN   \n",
       "\n",
       "    Unnamed: 9  ... Unnamed: 230 Unnamed: 231 Unnamed: 232 Unnamed: 233  \\\n",
       "0          NaN  ...          NaN          NaN          NaN          NaN   \n",
       "1       2/8/13  ...          NaN          NaN          NaN          NaN   \n",
       "2          16S  ...          NaN          NaN          NaN          NaN   \n",
       "3        98.15  ...          NaN          NaN          NaN          NaN   \n",
       "4          NaN  ...          NaN          NaN          NaN          NaN   \n",
       "..         ...  ...          ...          ...          ...          ...   \n",
       "281        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "282        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "283        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "284        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "285        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "\n",
       "    Unnamed: 234 Unnamed: 235 Unnamed: 236 Unnamed: 237 Unnamed: 238  \\\n",
       "0            NaN          NaN          NaN          NaN          NaN   \n",
       "1            NaN          NaN          NaN          NaN          NaN   \n",
       "2            NaN          NaN          NaN          NaN          NaN   \n",
       "3            NaN          NaN          NaN          NaN          NaN   \n",
       "4            NaN          NaN          NaN          NaN          NaN   \n",
       "..           ...          ...          ...          ...          ...   \n",
       "281          NaN          NaN          NaN          NaN          NaN   \n",
       "282          NaN          NaN          NaN          NaN          NaN   \n",
       "283          NaN          NaN          NaN          NaN          NaN   \n",
       "284          NaN          NaN          NaN          NaN          NaN   \n",
       "285          NaN          NaN          NaN          NaN          NaN   \n",
       "\n",
       "    Unnamed: 239  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "..           ...  \n",
       "281          NaN  \n",
       "282          NaN  \n",
       "283          NaN  \n",
       "284          NaN  \n",
       "285          NaN  \n",
       "\n",
       "[286 rows x 240 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_microbiome = pd.read_csv(fp_microbiome)\n",
    "df_microbiome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7905d5b2",
   "metadata": {},
   "source": [
    "## Create Summary Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2dd5bd8",
   "metadata": {},
   "source": [
    "### Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "29afb966",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_row_values(wanted_label: str):\n",
    "    target = clean_label(wanted_label)\n",
    "    idx = labels_clean[labels_clean == target].index\n",
    "    if len(idx) == 0:\n",
    "        return pd.Series(index=col_date, dtype=float)\n",
    "    return df_microbiome.loc[idx[0], col_date]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "c252b182",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_row_values(wanted_label: str):\n",
    "    \"\"\"\n",
    "    Return a Series indexed by your existing `col_date` labels, pulling values\n",
    "    from the row whose left label == wanted_label. Works even if the displayed\n",
    "    dates live in a header row and the actual DataFrame column names differ.\n",
    "    \"\"\"\n",
    "    target = clean_label(wanted_label)\n",
    "    idx = labels_clean[labels_clean == target].index\n",
    "    if len(idx) == 0:\n",
    "        return pd.Series(index=col_date, dtype=float)\n",
    "\n",
    "    # use your already-detected header row `i` if present; otherwise find it\n",
    "    try:\n",
    "        header_row_idx = i\n",
    "    except NameError:\n",
    "        _hdr = labels_clean[labels_clean == \"Date stool sample collected\"].index\n",
    "        header_row_idx = _hdr[0] if len(_hdr) else idx[0]\n",
    "\n",
    "    # map displayed date labels -> physical DataFrame column names\n",
    "    date_set = set(col_date)\n",
    "    date_to_col = {}\n",
    "    for c in cols:\n",
    "        v = clean_label(df_microbiome.at[header_row_idx, c])\n",
    "        if v in date_set and v not in date_to_col:\n",
    "            date_to_col[v] = c\n",
    "\n",
    "    r = idx[0]\n",
    "    # build a Series indexed by your `col_date` (return NaN where unmapped)\n",
    "    return pd.Series(\n",
    "        [df_microbiome.at[r, date_to_col[d]] if d in date_to_col else None for d in col_date],\n",
    "        index=col_date\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "a7bb2af9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LS Gut Microbiome</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Unnamed: 3</th>\n",
       "      <th>Unnamed: 4</th>\n",
       "      <th>Unnamed: 5</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 230</th>\n",
       "      <th>Unnamed: 231</th>\n",
       "      <th>Unnamed: 232</th>\n",
       "      <th>Unnamed: 233</th>\n",
       "      <th>Unnamed: 234</th>\n",
       "      <th>Unnamed: 235</th>\n",
       "      <th>Unnamed: 236</th>\n",
       "      <th>Unnamed: 237</th>\n",
       "      <th>Unnamed: 238</th>\n",
       "      <th>Unnamed: 239</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sequenced in Knight Lab from LS Frozen Stool S...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Date stool sample collected</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Max</td>\n",
       "      <td>Average</td>\n",
       "      <td>12/28/11</td>\n",
       "      <td>4/3/12</td>\n",
       "      <td>8/7/12</td>\n",
       "      <td>11/6/12</td>\n",
       "      <td>1/26/13</td>\n",
       "      <td>2/8/13</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16S Relative Abundance</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>16S</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sum of 16S Relative Abundances</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.15</td>\n",
       "      <td>97.89</td>\n",
       "      <td>97.98</td>\n",
       "      <td>98.39</td>\n",
       "      <td>97.89</td>\n",
       "      <td>98.15</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Dominate species from Metagenomics comparison</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>Escherichia coli Max HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.6032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>Morganella morganii Ave HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000108823529411765</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>Morganella morganii Max HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>Bifidobacterium longum Ave HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.108320588235294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>Bifidobacterium longum Max HE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.2247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>286 rows × 240 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    LS Gut Microbiome   \\\n",
       "0    Sequenced in Knight Lab from LS Frozen Stool S...   \n",
       "1                          Date stool sample collected   \n",
       "2                               16S Relative Abundance   \n",
       "3                       Sum of 16S Relative Abundances   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "281                            Escherichia coli Max HE   \n",
       "282                         Morganella morganii Ave HE   \n",
       "283                         Morganella morganii Max HE   \n",
       "284                      Bifidobacterium longum Ave HE   \n",
       "285                      Bifidobacterium longum Max HE   \n",
       "\n",
       "                                        Unnamed: 1 Unnamed: 2 Unnamed: 3  \\\n",
       "0                                              NaN        NaN        NaN   \n",
       "1                                              NaN        Max    Average   \n",
       "2                                              NaN        NaN        NaN   \n",
       "3                                              NaN        NaN        NaN   \n",
       "4    Dominate species from Metagenomics comparison        NaN        NaN   \n",
       "..                                             ...        ...        ...   \n",
       "281                                            NaN        NaN        NaN   \n",
       "282                                            NaN        NaN        NaN   \n",
       "283                                            NaN        NaN        NaN   \n",
       "284                                            NaN        NaN        NaN   \n",
       "285                                            NaN        NaN        NaN   \n",
       "\n",
       "               Unnamed: 4 Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8  \\\n",
       "0                     NaN        NaN        NaN        NaN        NaN   \n",
       "1                12/28/11     4/3/12     8/7/12    11/6/12    1/26/13   \n",
       "2                     16S        16S        16S        16S        16S   \n",
       "3                   97.15      97.89      97.98      98.39      97.89   \n",
       "4                     NaN        NaN        NaN        NaN        NaN   \n",
       "..                    ...        ...        ...        ...        ...   \n",
       "281                0.6032        NaN        NaN        NaN        NaN   \n",
       "282  0.000108823529411765        NaN        NaN        NaN        NaN   \n",
       "283                0.0004        NaN        NaN        NaN        NaN   \n",
       "284     0.108320588235294        NaN        NaN        NaN        NaN   \n",
       "285                1.2247        NaN        NaN        NaN        NaN   \n",
       "\n",
       "    Unnamed: 9  ... Unnamed: 230 Unnamed: 231 Unnamed: 232 Unnamed: 233  \\\n",
       "0          NaN  ...          NaN          NaN          NaN          NaN   \n",
       "1       2/8/13  ...          NaN          NaN          NaN          NaN   \n",
       "2          16S  ...          NaN          NaN          NaN          NaN   \n",
       "3        98.15  ...          NaN          NaN          NaN          NaN   \n",
       "4          NaN  ...          NaN          NaN          NaN          NaN   \n",
       "..         ...  ...          ...          ...          ...          ...   \n",
       "281        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "282        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "283        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "284        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "285        NaN  ...          NaN          NaN          NaN          NaN   \n",
       "\n",
       "    Unnamed: 234 Unnamed: 235 Unnamed: 236 Unnamed: 237 Unnamed: 238  \\\n",
       "0            NaN          NaN          NaN          NaN          NaN   \n",
       "1            NaN          NaN          NaN          NaN          NaN   \n",
       "2            NaN          NaN          NaN          NaN          NaN   \n",
       "3            NaN          NaN          NaN          NaN          NaN   \n",
       "4            NaN          NaN          NaN          NaN          NaN   \n",
       "..           ...          ...          ...          ...          ...   \n",
       "281          NaN          NaN          NaN          NaN          NaN   \n",
       "282          NaN          NaN          NaN          NaN          NaN   \n",
       "283          NaN          NaN          NaN          NaN          NaN   \n",
       "284          NaN          NaN          NaN          NaN          NaN   \n",
       "285          NaN          NaN          NaN          NaN          NaN   \n",
       "\n",
       "    Unnamed: 239  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "..           ...  \n",
       "281          NaN  \n",
       "282          NaN  \n",
       "283          NaN  \n",
       "284          NaN  \n",
       "285          NaN  \n",
       "\n",
       "[286 rows x 240 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_microbiome = pd.read_csv(fp_microbiome)\n",
    "df_microbiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "896c33a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Sum16S</th>\n",
       "      <th>SumMetagenomics</th>\n",
       "      <th>Shannon16S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>97.15</td>\n",
       "      <td>99.957</td>\n",
       "      <td>5.444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4/3/12</td>\n",
       "      <td>97.89</td>\n",
       "      <td>99.985</td>\n",
       "      <td>4.633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8/7/12</td>\n",
       "      <td>97.98</td>\n",
       "      <td>99.965</td>\n",
       "      <td>5.997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/6/12</td>\n",
       "      <td>98.39</td>\n",
       "      <td>99.983</td>\n",
       "      <td>5.244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1/26/13</td>\n",
       "      <td>97.89</td>\n",
       "      <td>99.987</td>\n",
       "      <td>6.235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>6/16/19</td>\n",
       "      <td>99.06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>6/23/19</td>\n",
       "      <td>99.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>7/14/19</td>\n",
       "      <td>99.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>7/28/19</td>\n",
       "      <td>98.52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>99.32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.122</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>159 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Date  Sum16S  SumMetagenomics  Shannon16S\n",
       "0    12/28/11   97.15           99.957       5.444\n",
       "1      4/3/12   97.89           99.985       4.633\n",
       "2      8/7/12   97.98           99.965       5.997\n",
       "3     11/6/12   98.39           99.983       5.244\n",
       "4     1/26/13   97.89           99.987       6.235\n",
       "..        ...     ...              ...         ...\n",
       "154   6/16/19   99.06              NaN       6.082\n",
       "155   6/23/19   99.10              NaN       5.596\n",
       "156   7/14/19   99.00              NaN       6.007\n",
       "157   7/28/19   98.52              NaN       6.133\n",
       "158   8/11/19   99.32              NaN       6.122\n",
       "\n",
       "[159 rows x 4 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = list(df_microbiome.columns) # all columns\n",
    "col_label = find_exact(\"Date stool sample collected\") or cols[0] # left label column (use your finder)\n",
    "\n",
    "# locate the row whose label is exactly \"Date stool sample collected\"\n",
    "labels_clean = df_microbiome[col_label].map(clean_label)\n",
    "i = labels_clean[labels_clean == \"Date stool sample collected\"].index[0]\n",
    "\n",
    "# grab the whole row, then drop the non-date entries\n",
    "row_vals = [clean_label(v) for v in df_microbiome.loc[i, cols].tolist()]\n",
    "col_date = [    \n",
    "            v for v in row_vals\n",
    "            if v not in (None, \"Date stool sample collected\", \"Max\", \"Average\")\n",
    "]\n",
    "\n",
    "# ensure uniqueness while preserving the original order\n",
    "col_date = uniq_order(col_date)\n",
    "\n",
    "# pull exactly the three rows you asked for\n",
    "sum16s   = pd.to_numeric(get_row_values(\"Sum of 16S Relative Abundances\"), errors=\"coerce\")\n",
    "sum_meta = pd.to_numeric(get_row_values(\"Sum of Metagenomics Relative Abundances\"), errors=\"coerce\")\n",
    "shannon  = pd.to_numeric(get_row_values(\"Shannon Diversity Index Score Computed from 16S\"), errors=\"coerce\")\n",
    "\n",
    "# assemble the summary strictly from your date_cols (no extra logic)\n",
    "df_microbiomeSummary = (\n",
    "                          pd.DataFrame(index=col_date)\n",
    "                            .assign(Sum16S=sum16s)\n",
    "                            .assign(SumMetagenomics=sum_meta)\n",
    "                            .assign(Shannon16S=shannon)\n",
    "                            .rename_axis(\"Date\")\n",
    "                            .reset_index()\n",
    "                            .dropna(how=\"all\", subset=[\"Sum16S\", \"SumMetagenomics\", \"Shannon16S\"])\n",
    ")\n",
    "\n",
    "df_microbiomeSummary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "2a605004",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out the summary\n",
    "df_microbiomeSummary.to_csv(fp_microbiomeSummary, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18f523c6",
   "metadata": {},
   "source": [
    "## Create Long Format"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcbe9a76",
   "metadata": {},
   "source": [
    "### Mapping Dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "4fffeaed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Species to Panel mapping\n",
    "dict_speciesToPanels = {\n",
    "\n",
    "                            # 16S\n",
    "                            \"p__Proteobacteria;cf__Enterobacteriaceae;g__Morganell\": \"16S\",  \n",
    "                            \"p__Proteobacteria;cf__Enterobacteriaceae;g__Morganella\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Enterobacteriaceae;__\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Enterobacteriaceae;g__\": \"16S\",\n",
    "                            \"p__Bacteroidetes;f__Bacteroidaceae;g__Bacteroides\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Ruminococcaceae;g__\": \"16S\",\n",
    "                            \"p__Verrucomicrobia;f__Verrucomicrobiaceae;g__Akkermansia\": \"16S\",\n",
    "                            \"p__Firmicutes;f__;g__\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Lachnospiraceae;g__Coprococcus\": \"16S\",\n",
    "                            \"p__Bacteroidetes;f__Prevotellaceae;g__Prevotella\": \"16S\",\n",
    "                            \"p__Bacteroidetes;f__Rikenellaceae;g__\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Ruminococcaceae;g__Faecalibacterium\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Lachnospiraceae;g__\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Lachnospiraceae;g__Lachnospira\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Lachnospiraceae;g__Blautia\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Christensenellaceae;g__\": \"16S\",\n",
    "                            \"p__Euryarchaeota;f__Methanobacteriaceae;g__Methanobrevibacter\": \"16S\",\n",
    "                            \"p__Bacteroidetes;f__Porphyromonadaceae;g__Parabacteroides\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Enterobacteriaceae;g__Providencia\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Ruminococcaceae;g__Ruminococcus\": \"16S\",\n",
    "                            \"p__Bacteroidetes;f__[Paraprevotellaceae];g__[Prevotella]\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Lachnospiraceae;g__[Ruminococcus]\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Enterobacteriaceae;g__Serratia\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Veillonellaceae;g__Dialister\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Lachnospiraceae;g__Dorea\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Lachnospiraceae;g__Roseburia\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Erysipelotrichaceae;g__\": \"16S\",\n",
    "                            \"p__Bacteroidetes;f__S24-7;g__\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Pseudomonadaceae;g__Pseudomonas\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__;g__\": \"16S\",\n",
    "                            \"p__Fusobacteria;f__Fusobacteriaceae;g__Fusobacterium\": \"16S\",\n",
    "                            \"p__Bacteroidetes;f__[Barnesiellaceae];g__\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Veillonellaceae;g__Phascolarctobacterium\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Enterobacteriaceae;g__Erwinia\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Alcaligenaceae;g__Sutterella\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Desulfovibrionaceae;g__Bilophila\": \"16S\",\n",
    "                            \"p__Cyanobacteria;f__;g__\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Ruminococcaceae;g__Oscillospira\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Erysipelotrichaceae;g__[Eubacterium]\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Desulfovibrionaceae;g__Desulfovibrio\": \"16S\",\n",
    "                            \"p__Tenericutes;f__;g__\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Burkholderiaceae;g__\": \"16S\",\n",
    "                            \"p__Synergistetes;f__Synergistaceae;g__Synergistes\": \"16S\",\n",
    "                            \"p__Firmicutes;f__[Tissierellaceae];g__Parvimonas\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Clostridiaceae;g__\": \"16S\",\n",
    "                            \"p__Actinobacteria;f__Coriobacteriaceae;g__\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Clostridiaceae;g__Clostridium\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Veillonellaceae;g__Veillonella\": \"16S\",\n",
    "                            \"p__Proteobacteria;f__Pasteurellaceae;g__Aggregatibacter\": \"16S\",\n",
    "                            \"p__Actinobacteria;f__Bifidobacteriaceae;g__Bifidobacterium\": \"16S\",\n",
    "                            \"p__Firmicutes;f__Lachnospiraceae;g__Lachnobacterium\": \"16S\",\n",
    "\n",
    "                            # Xin Fang MetaphlAN 2 Metagenomics \n",
    "                            \"Sum of Metagenomics Relative Abundances\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Escherichia_coli\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Akkermansia_muciniphila\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Faecalibacterium_prausnitzii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eubacterium_rectale\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Methanobrevibacter_smithii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Subdoligranulum_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_uniformis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_vulgatus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eubacterium_eligens\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_intestinalis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Morganella_morganii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eubacterium_siraeum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_cellulosilyticus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Dialister_invisus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Escherichia_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Citrobacter_freundii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_sp_L2_50\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_pectinophilus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_caccae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Ruminococcus_lactaris\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Alistipes_putredinis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Parabacteroides_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Roseburia_inulinivorans\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Enterobacteria_phage_HK97\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_nexile\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lachnospiraceae_bacterium_1_1_57FAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Ruminococcus_bromii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Klebsiella_pneumoniae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Parabacteroides_merdae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Sutterella_wadsworthensis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Barnesiella_intestinihominis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"C2likevirus_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Ruminococcus_torques\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroidales_bacterium_ph8\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_thetaiotaomicron\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_ovatus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Roseburia_intestinalis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Roseburia_hominis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Providencia_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Enterobacteriaceae_bacterium_9_2_54FAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_salyersiae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_fragilis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Dorea_longicatena\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Alistipes_sp_HGB5\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Ruminococcus_sp_5_1_39BFAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Parvimonas_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Alistipes_shahii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Ruminococcus_obeum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eubacterium_hallii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Coprococcus_sp_ART55_1\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bifidobacterium_longum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Alistipes_onderdonkii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Fusobacterium_nucleatum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"T5likevirus_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Odoribacter_splanchnicus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_dorei\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Coprococcus_comes\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Oscillibacter_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bilophila_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Collinsella_aerofaciens\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Klebsiella_oxytoca\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Streptococcus_thermophilus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Dorea_formicigenerans\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eubacterium_ramulus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lachnospiraceae_bacterium_3_1_46FAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lachnospiraceae_bacterium_3_1_57FAA_CT1\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Alistipes_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Alistipes_sp_AP11\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Subdoligranulum_sp_4_3_54A2FAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Klebsiella_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Coprobacillus_sp_29_1\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lachnospiraceae_bacterium_5_1_63FAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Methanosphaera_stadtmanae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Roseburia_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_hathewayi\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Aggregatibacter_segnis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Enterobacter_cloacae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Anaerostipes_hadrus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Peptostreptococcus_stomatis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Desulfovibrio_piger\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Parabacteroides_goldsteinii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_symbiosum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Coprococcus_catus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bifidobacterium_animalis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Parvimonas_micra\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Anaerotruncus_colihominis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Porphyromonas_asaccharolytica\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Oxalobacter_formigenes\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Citrobacter_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_bolteae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Veillonella_parvula\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_xylanisolvens\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Gemella_morbillorum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Solobacterium_moorei\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Alistipes_indistinctus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Streptococcus_salivarius\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Erysipelotrichaceae_bacterium_2_2_44A\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Streptococcus_parasanguinis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Pseudomonas_phage_F116\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_leptum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lachnospiraceae_bacterium_8_1_57FAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lachnospiraceae_bacterium_7_1_58FAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Holdemania_filiformis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Veillonella_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bilophila_wadsworthia\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Burkholderia_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lactobacillus_casei_paracasei\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Fusobacterium_ulcerans\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Adlercreutzia_equolifaciens\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Alistipes_senegalensis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Aggregatibacter_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Erysipelotrichaceae_bacterium_6_1_45\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Pseudomonas_aeruginosa\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Burkholderiales_bacterium_1_1_47\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eggerthella_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Anaerotruncus_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eubacterium_ventriosum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridiaceae_bacterium_JC118\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Ruminococcaceae_bacterium_D16\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Anaerococcus_vaginalis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Haemophilus_parainfluenzae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Methanobrevibacter_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eikenella_corrodens\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eubacterium_infirmum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Streptococcus_anginosus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Burkholderia_cenocepacia\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_asparagiforme\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Granulicatella_adiacens\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_scindens\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Coprobacter_fastidiosus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Alistipes_finegoldii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_nordii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Streptococcus_australis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Ruminococcus_callidus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Hafnia_alvei\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bifidobacterium_breve\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Parasutterella_excrementihominis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lactococcus_phage_P680\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Veillonella_atypica\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lactococcus_lactis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Pseudomonas_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lactobacillus_helveticus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridiales_bacterium_1_7_47FAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lactobacillus_acidophilus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Blautia_hydrogenotrophica\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Aeromonas_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Pantoea_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_citroniae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Dialister_succinatiphilus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Peptoniphilus_harei\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Parabacteroides_distasonis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Gordonibacter_pamelaeae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Fusobacterium_periodonticum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Shigella_sonnei\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Flavonifractor_plautii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eggerthella_lenta\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Marvinbryantia_formatexigens\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Erysipelotrichaceae_bacterium_21_3\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Bacteroides_faecis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_innocuum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Peptoniphilus_lacrimalis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Clostridium_bartlettii\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Campylobacter_showae\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Rothia_mucilaginosa\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Peptostreptococcaceae_noname_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Enterococcus_faecium\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Pediococcus_acidilactici\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Desulfovibrio_desulfuricans\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Leuconostoc_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Streptococcus_vestibularis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Lachnospiraceae_bacterium_5_1_57FAA\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Streptococcus_constellatus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Torque_teno_virus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Holdemania_unclassified\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Pseudoflavonifractor_capillosus\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Prevotella_timonensis\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Eubacterium_limosum\": \"Xin Fang MetaphlAN 2 Metagenomics\",\n",
    "                            \"Oscillibacter_sp_KLE_1745\": \"Xin Fang MetaphlAN 2 Metagenomics\"\n",
    "                                                        \n",
    "}\n",
    "\n",
    "# Map Species to Dominant Species for 16S only (some species map to None)\n",
    "dict_speciesToDominantSpecies16S = {\n",
    "                                    \"p__Proteobacteria;cf__Enterobacteriaceae;g__Morganella\": \"Morganella_morganii\",\n",
    "                                    \"p__Proteobacteria;f__Enterobacteriaceae;__\": \"Klebsiella_pneumoniae\",\n",
    "                                    \"p__Proteobacteria;f__Enterobacteriaceae;g__\": \"Escherichia_coli\",\n",
    "                                    \"p__Bacteroidetes;f__Bacteroidaceae;g__Bacteroides\": \"Bacteroides_uniformis\",\n",
    "                                    \"p__Firmicutes;f__Ruminococcaceae;g__\": \"Subdoligranulum_unclassified\",\n",
    "                                    \"p__Verrucomicrobia;f__Verrucomicrobiaceae;g__Akkermansia\": \"Akkermansia_muciniphila\",\n",
    "                                    \"p__Firmicutes;f__;g__\": \"Clostridium_sp_L2_50\",\n",
    "                                    \"p__Firmicutes;f__Lachnospiraceae;g__Coprococcus\": None,\n",
    "                                    \"p__Bacteroidetes;f__Prevotellaceae;g__Prevotella\": None,\n",
    "                                    \"p__Bacteroidetes;f__Rikenellaceae;g__\": \"Alistipes_putredinis\",\n",
    "                                    \"p__Firmicutes;f__Ruminococcaceae;g__Faecalibacterium\": \"Faecalibacterium_prausnitzii\",\n",
    "                                    \"p__Firmicutes;f__Lachnospiraceae;g__\": None,\n",
    "                                    \"p__Firmicutes;f__Lachnospiraceae;g__Lachnospira\": None,\n",
    "                                    \"p__Firmicutes;f__Lachnospiraceae;g__Blautia\": None,\n",
    "                                    \"p__Firmicutes;f__Christensenellaceae;g__\": None,\n",
    "                                    \"p__Euryarchaeota;f__Methanobacteriaceae;g__Methanobrevibacter\": \"Methanobrevibacter_smithii\",\n",
    "                                    \"p__Bacteroidetes;f__Porphyromonadaceae;g__Parabacteroides\": None,\n",
    "                                    \"p__Proteobacteria;f__Enterobacteriaceae;g__Providencia\": None,\n",
    "                                    \"p__Firmicutes;f__Ruminococcaceae;g__Ruminococcus\": None,\n",
    "                                    \"p__Bacteroidetes;f__[Paraprevotellaceae];g__[Prevotella]\": None,\n",
    "                                    \"p__Firmicutes;f__Lachnospiraceae;g__[Ruminococcus]\": None,\n",
    "                                    \"p__Proteobacteria;f__Enterobacteriaceae;g__Serratia\": None,\n",
    "                                    \"p__Firmicutes;f__Veillonellaceae;g__Dialister\": None,\n",
    "                                    \"p__Firmicutes;f__Lachnospiraceae;g__Dorea\": None,\n",
    "                                    \"p__Firmicutes;f__Lachnospiraceae;g__Roseburia\": None,\n",
    "                                    \"p__Firmicutes;f__Erysipelotrichaceae;g__\": None,\n",
    "                                    \"p__Bacteroidetes;f__S24-7;g__\": None,\n",
    "                                    \"p__Proteobacteria;f__Pseudomonadaceae;g__Pseudomonas\": None,\n",
    "                                    \"p__Proteobacteria;f__;g__\": None,\n",
    "                                    \"p__Fusobacteria;f__Fusobacteriaceae;g__Fusobacterium\": None,\n",
    "                                    \"p__Bacteroidetes;f__[Barnesiellaceae];g__\": None,\n",
    "                                    \"p__Firmicutes;f__Veillonellaceae;g__Phascolarctobacterium\": None,\n",
    "                                    \"p__Proteobacteria;f__Enterobacteriaceae;g__Erwinia\": None,\n",
    "                                    \"p__Proteobacteria;f__Alcaligenaceae;g__Sutterella\": None,\n",
    "                                    \"p__Proteobacteria;f__Desulfovibrionaceae;g__Bilophila\": None,\n",
    "                                    \"p__Cyanobacteria;f__;g__\": None,\n",
    "                                    \"p__Firmicutes;f__Ruminococcaceae;g__Oscillospira\": None,\n",
    "                                    \"p__Firmicutes;f__Erysipelotrichaceae;g__[Eubacterium]\": None,\n",
    "                                    \"p__Proteobacteria;f__Desulfovibrionaceae;g__Desulfovibrio\": None,\n",
    "                                    \"p__Tenericutes;f__;g__\": None,\n",
    "                                    \"p__Proteobacteria;f__Burkholderiaceae;g__\": None,\n",
    "                                    \"p__Synergistetes;f__Synergistaceae;g__Synergistes\": None,\n",
    "                                    \"p__Firmicutes;f__[Tissierellaceae];g__Parvimonas\": None,\n",
    "                                    \"p__Firmicutes;f__Clostridiaceae;g__\": None,\n",
    "                                    \"p__Actinobacteria;f__Coriobacteriaceae;g__\": None,\n",
    "                                    \"p__Firmicutes;f__Clostridiaceae;g__Clostridium\": None,\n",
    "                                    \"p__Firmicutes;f__Veillonellaceae;g__Veillonella\": None,\n",
    "                                    \"p__Proteobacteria;f__Pasteurellaceae;g__Aggregatibacter\": None,\n",
    "                                    \"p__Actinobacteria;f__Bifidobacteriaceae;g__Bifidobacterium\": None,\n",
    "                                    \"p__Firmicutes;f__Lachnospiraceae;g__Lachnobacterium\": None\n",
    "}\n",
    "\n",
    "dict_speciesToDominantXF = {\n",
    "                                    \"Xin Fang MetaphlAN 2 Metagenomics\" : None\n",
    "}\n",
    "\n",
    "# Healthy reference (Average)\n",
    "dict_speciesToHealthyAverage = {\n",
    "                                \"Bacteroides_vulgatus\": 6.05498529411765,\n",
    "                                \"Bacteroides_ovatus\": 4.23238823529412,\n",
    "                                \"Prevotella_copri\": 3.30236764705882,\n",
    "                                \"Eubacterium_rectale\": 3.05285588235294,\n",
    "                                \"Faecalibacterium_prausnitzii\": 2.95155,\n",
    "                                \"Dialister_invisus\": 2.16144117647059,\n",
    "                                \"Bacteroides_dorei\": 2.11637352941176,\n",
    "                                \"Ruminococcus_bromii\": 1.85948235294118,\n",
    "                                \"Bacteroides_uniformis\": 1.7114,\n",
    "                                \"Bacteroides_cellulosilyticus\": 1.48938235294118,\n",
    "                                \"Alistipes_finegoldii\": 1.15306764705882,\n",
    "                                \"Akkermansia_muciniphila\": 0.984429411764706,\n",
    "                                \"Eubacterium_siraeum\": 0.966767647058824,\n",
    "                                \"Eubacterium_eligens\": 0.939023529411765,\n",
    "                                \"Bacteroides_fragilis\": 0.905426470588236,\n",
    "                                \"Methanobrevibacter_smithii\": 0.111332352941176,\n",
    "                                \"Parvimonas_micra\": 0.0172323529411765,\n",
    "                                \"Streptococcus_thermophilus\": 0.0225911764705882,\n",
    "                                \"Bacteroides_intestinalis\": 0.630638235294118,\n",
    "                                \"Escherichia_coli\": 0.0551147058823529,\n",
    "                                \"Morganella_morganii\": 0.000108823529411765,\n",
    "                                \"Bifidobacterium_longum\": 0.108320588235294\n",
    "}\n",
    "\n",
    "# Healthy reference (Max)\n",
    "dict_speciesToHealthyMax = {\n",
    "                            \"Bacteroides_vulgatus\": 26.5957,\n",
    "                            \"Bacteroides_ovatus\": 41.7481,\n",
    "                            \"Prevotella_copri\": 46.0315,\n",
    "                            \"Eubacterium_rectale\": 21.8571,\n",
    "                            \"Faecalibacterium_prausnitzii\": 11.7871,\n",
    "                            \"Dialister_invisus\": 27.0459,\n",
    "                            \"Bacteroides_dorei\": 9.7387,\n",
    "                            \"Ruminococcus_bromii\": 9.6994,\n",
    "                            \"Bacteroides_uniformis\": 4.2107,\n",
    "                            \"Bacteroides_cellulosilyticus\": 18.1811,\n",
    "                            \"Alistipes_finegoldii\": 10.7227,\n",
    "                            \"Akkermansia_muciniphila\": 8.439,\n",
    "                            \"Eubacterium_siraeum\": 12.1935,\n",
    "                            \"Eubacterium_eligens\": 4.6383,\n",
    "                            \"Bacteroides_fragilis\": 4.1907,\n",
    "                            \"Methanobrevibacter_smithii\": 1.295,\n",
    "                            \"Parvimonas_micra\": 0.0718,\n",
    "                            \"Streptococcus_thermophilus\": 0.1446,\n",
    "                            \"Bacteroides_intestinalis\": 6.4841,\n",
    "                            \"Escherichia_coli\": 0.6032,\n",
    "                            \"Morganella_morganii\": 0.0004,\n",
    "                            \"Bifidobacterium_longum\": 1.2247\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "438bef52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Biomarker</th>\n",
       "      <th>Value</th>\n",
       "      <th>Species</th>\n",
       "      <th>Abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10/6/1999</td>\n",
       "      <td>Weight</td>\n",
       "      <td>202.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5/7/2001</td>\n",
       "      <td>Weight</td>\n",
       "      <td>201.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11/13/2003</td>\n",
       "      <td>Weight</td>\n",
       "      <td>192.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10/4/2005</td>\n",
       "      <td>Weight</td>\n",
       "      <td>179.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4/17/2007</td>\n",
       "      <td>Weight</td>\n",
       "      <td>179.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38682</th>\n",
       "      <td>7/11/15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oscillibacter_sp_KLE_1745</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38683</th>\n",
       "      <td>7/12/15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oscillibacter_sp_KLE_1745</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38684</th>\n",
       "      <td>7/15/15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oscillibacter_sp_KLE_1745</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38685</th>\n",
       "      <td>7/19/15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oscillibacter_sp_KLE_1745</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38686</th>\n",
       "      <td>7/26/15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oscillibacter_sp_KLE_1745</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38687 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Date Biomarker  Value                    Species  Abundance\n",
       "0       10/6/1999    Weight  202.0                        NaN        NaN\n",
       "1        5/7/2001    Weight  201.0                        NaN        NaN\n",
       "2      11/13/2003    Weight  192.0                        NaN        NaN\n",
       "3       10/4/2005    Weight  179.0                        NaN        NaN\n",
       "4       4/17/2007    Weight  179.0                        NaN        NaN\n",
       "...           ...       ...    ...                        ...        ...\n",
       "38682     7/11/15       NaN    NaN  Oscillibacter_sp_KLE_1745        0.0\n",
       "38683     7/12/15       NaN    NaN  Oscillibacter_sp_KLE_1745        0.0\n",
       "38684     7/15/15       NaN    NaN  Oscillibacter_sp_KLE_1745        0.0\n",
       "38685     7/19/15       NaN    NaN  Oscillibacter_sp_KLE_1745        0.0\n",
       "38686     7/26/15       NaN    NaN  Oscillibacter_sp_KLE_1745        0.0\n",
       "\n",
       "[38687 rows x 5 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create long format\n",
    "# Species, Panel, DominantSpecies, HealthyAverage, HealthyMax, Abundance\n",
    "lst_species = dict_speciesToPanels.keys()\n",
    "\n",
    "# Mathch each entry for abundance to speciies and date\n",
    "for sp in lst_species:\n",
    "    s = pd.to_numeric(get_row_values(sp), errors=\"coerce\")  # indexed by your existing col_date\n",
    "    for date in col_date:\n",
    "        val = s.get(date)\n",
    "        if pd.notna(val):\n",
    "            records.append({\"Date\": date, \"Species\": sp, \"Abundance\": val})\n",
    "\n",
    "df_longMicrobiome = pd.DataFrame.from_records(records)\n",
    "df_longMicrobiome \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "92153cbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>Panel</th>\n",
       "      <th>DominantSpecies</th>\n",
       "      <th>HealthyAverage</th>\n",
       "      <th>HealthyMax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Adlercreutzia_equolifaciens</td>\n",
       "      <td>0.031</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Aeromonas_unclassified</td>\n",
       "      <td>0.000</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Aggregatibacter_segnis</td>\n",
       "      <td>0.000</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Aggregatibacter_unclassified</td>\n",
       "      <td>0.000</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>12.328</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.984429</td>\n",
       "      <td>8.439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16138</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Proteobacteria;f__Pasteurellaceae;g__Aggreg...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16139</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Proteobacteria;f__Pseudomonadaceae;g__Pseud...</td>\n",
       "      <td>0.060</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16140</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Synergistetes;f__Synergistaceae;g__Synergistes</td>\n",
       "      <td>0.030</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16141</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Tenericutes;f__;g__</td>\n",
       "      <td>0.580</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16142</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Verrucomicrobia;f__Verrucomicrobiaceae;g__A...</td>\n",
       "      <td>1.600</td>\n",
       "      <td>16S</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16143 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Date                                            Species  Abundance  \\\n",
       "0      12/28/11                        Adlercreutzia_equolifaciens      0.031   \n",
       "1      12/28/11                             Aeromonas_unclassified      0.000   \n",
       "2      12/28/11                             Aggregatibacter_segnis      0.000   \n",
       "3      12/28/11                       Aggregatibacter_unclassified      0.000   \n",
       "4      12/28/11                            Akkermansia_muciniphila     12.328   \n",
       "...         ...                                                ...        ...   \n",
       "16138   8/11/19  p__Proteobacteria;f__Pasteurellaceae;g__Aggreg...      0.000   \n",
       "16139   8/11/19  p__Proteobacteria;f__Pseudomonadaceae;g__Pseud...      0.060   \n",
       "16140   8/11/19  p__Synergistetes;f__Synergistaceae;g__Synergistes      0.030   \n",
       "16141   8/11/19                             p__Tenericutes;f__;g__      0.580   \n",
       "16142   8/11/19  p__Verrucomicrobia;f__Verrucomicrobiaceae;g__A...      1.600   \n",
       "\n",
       "                                   Panel          DominantSpecies  \\\n",
       "0      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "1      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "2      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "3      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "4      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "...                                  ...                      ...   \n",
       "16138                                16S                     None   \n",
       "16139                                16S                     None   \n",
       "16140                                16S                     None   \n",
       "16141                                16S                     None   \n",
       "16142                                16S  Akkermansia_muciniphila   \n",
       "\n",
       "       HealthyAverage  HealthyMax  \n",
       "0                 NaN         NaN  \n",
       "1                 NaN         NaN  \n",
       "2                 NaN         NaN  \n",
       "3                 NaN         NaN  \n",
       "4            0.984429       8.439  \n",
       "...               ...         ...  \n",
       "16138             NaN         NaN  \n",
       "16139             NaN         NaN  \n",
       "16140             NaN         NaN  \n",
       "16141             NaN         NaN  \n",
       "16142             NaN         NaN  \n",
       "\n",
       "[16143 rows x 7 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CLEAN start (do not reuse the old 'records')\n",
    "records = []\n",
    "\n",
    "# use the species actually present in this sheet (no recomputing anything else)\n",
    "species_present = [x for x in labels_clean.tolist()\n",
    "                   if (x is not None) and (x in dict_speciesToPanels)]\n",
    "\n",
    "# Date, Species, Abundance long format\n",
    "for sp in species_present:\n",
    "    s = pd.to_numeric(get_row_values(sp), errors=\"coerce\")  # indexed by your existing col_date\n",
    "    for d in col_date:\n",
    "        v = s.get(d)\n",
    "        if pd.notna(v):\n",
    "            records.append((d, sp, float(v)))\n",
    "\n",
    "df_longMicrobiome = pd.DataFrame(records, columns=[\"Date\", \"Species\", \"Abundance\"])\n",
    "\n",
    "# keep your original date order\n",
    "df_longMicrobiome[\"Date\"] = pd.Categorical(df_longMicrobiome[\"Date\"],\n",
    "                                           categories=col_date, ordered=True)\n",
    "df_longMicrobiome = df_longMicrobiome.sort_values([\"Date\", \"Species\"]).reset_index(drop=True)\n",
    "\n",
    "# map panels, dominant species, healthy average and max\n",
    "df_longMicrobiome[\"Panel\"] = df_longMicrobiome[\"Species\"].map(dict_speciesToPanels)\n",
    "df_longMicrobiome[\"DominantSpecies\"] = df_longMicrobiome[\"Species\"].map(dict_speciesToDominantSpecies16S)\n",
    "df_longMicrobiome[\"HealthyAverage\"] = df_longMicrobiome[\"Species\"].map(dict_speciesToHealthyAverage)\n",
    "df_longMicrobiome[\"HealthyMax\"] = df_longMicrobiome[\"Species\"].map(dict_speciesToHealthyMax)\n",
    "# END CLEAN\n",
    "df_longMicrobiome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "5a01b6bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>Panel</th>\n",
       "      <th>DominantSpecies</th>\n",
       "      <th>HealthyAverage</th>\n",
       "      <th>HealthyMax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Adlercreutzia_equolifaciens</td>\n",
       "      <td>0.031</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Aeromonas_unclassified</td>\n",
       "      <td>0.000</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Aggregatibacter_segnis</td>\n",
       "      <td>0.000</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Aggregatibacter_unclassified</td>\n",
       "      <td>0.000</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>12.328</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.984429</td>\n",
       "      <td>8.439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16138</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Proteobacteria;f__Pasteurellaceae;g__Aggreg...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16139</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Proteobacteria;f__Pseudomonadaceae;g__Pseud...</td>\n",
       "      <td>0.060</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16140</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Synergistetes;f__Synergistaceae;g__Synergistes</td>\n",
       "      <td>0.030</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16141</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Tenericutes;f__;g__</td>\n",
       "      <td>0.580</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16142</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Verrucomicrobia;f__Verrucomicrobiaceae;g__A...</td>\n",
       "      <td>1.600</td>\n",
       "      <td>16S</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16143 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Date                                            Species  Abundance  \\\n",
       "0      12/28/11                        Adlercreutzia_equolifaciens      0.031   \n",
       "1      12/28/11                             Aeromonas_unclassified      0.000   \n",
       "2      12/28/11                             Aggregatibacter_segnis      0.000   \n",
       "3      12/28/11                       Aggregatibacter_unclassified      0.000   \n",
       "4      12/28/11                            Akkermansia_muciniphila     12.328   \n",
       "...         ...                                                ...        ...   \n",
       "16138   8/11/19  p__Proteobacteria;f__Pasteurellaceae;g__Aggreg...      0.000   \n",
       "16139   8/11/19  p__Proteobacteria;f__Pseudomonadaceae;g__Pseud...      0.060   \n",
       "16140   8/11/19  p__Synergistetes;f__Synergistaceae;g__Synergistes      0.030   \n",
       "16141   8/11/19                             p__Tenericutes;f__;g__      0.580   \n",
       "16142   8/11/19  p__Verrucomicrobia;f__Verrucomicrobiaceae;g__A...      1.600   \n",
       "\n",
       "                                   Panel          DominantSpecies  \\\n",
       "0      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "1      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "2      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "3      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "4      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "...                                  ...                      ...   \n",
       "16138                                16S                     None   \n",
       "16139                                16S                     None   \n",
       "16140                                16S                     None   \n",
       "16141                                16S                     None   \n",
       "16142                                16S  Akkermansia_muciniphila   \n",
       "\n",
       "       HealthyAverage  HealthyMax  \n",
       "0                 NaN         NaN  \n",
       "1                 NaN         NaN  \n",
       "2                 NaN         NaN  \n",
       "3                 NaN         NaN  \n",
       "4            0.984429       8.439  \n",
       "...               ...         ...  \n",
       "16138             NaN         NaN  \n",
       "16139             NaN         NaN  \n",
       "16140             NaN         NaN  \n",
       "16141             NaN         NaN  \n",
       "16142             NaN         NaN  \n",
       "\n",
       "[16143 rows x 7 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# === LONG + MAPPINGS (one go, only using what you already defined) ===\n",
    "# uses: df_microbiome, labels_clean, col_date, get_row_values,\n",
    "#       dict_speciesToPanels, dict_speciesToDominantSpecies16S,\n",
    "#       dict_speciesToHealthyAverage, dict_speciesToHealthyMax\n",
    "\n",
    "# 1) long rows: (Date, Species, Abundance) — skip missing, keep zeros\n",
    "records = []\n",
    "species_present = [x for x in labels_clean.tolist()\n",
    "                   if (x is not None) and (x in dict_speciesToPanels)]\n",
    "\n",
    "for sp in species_present:\n",
    "    s = pd.to_numeric(get_row_values(sp), errors=\"coerce\")  # indexed by your existing col_date\n",
    "    for d in col_date:\n",
    "        v = s.get(d)\n",
    "        if pd.notna(v):\n",
    "            records.append((d, sp, float(v)))\n",
    "\n",
    "df_longMicrobiome = pd.DataFrame(records, columns=[\"Date\", \"Species\", \"Abundance\"])\n",
    "\n",
    "# 2) keep your original date order + sort\n",
    "df_longMicrobiome[\"Date\"] = pd.Categorical(df_longMicrobiome[\"Date\"],\n",
    "                                           categories=col_date, ordered=True)\n",
    "df_longMicrobiome = df_longMicrobiome.sort_values([\"Date\",\"Species\"]).reset_index(drop=True)\n",
    "\n",
    "# 3) add mappings (panel, dominant species, healthy refs)\n",
    "df_longMicrobiome[\"Panel\"]           = df_longMicrobiome[\"Species\"].map(dict_speciesToPanels)\n",
    "df_longMicrobiome[\"DominantSpecies\"] = df_longMicrobiome[\"Species\"].map(dict_speciesToDominantSpecies16S)\n",
    "df_longMicrobiome[\"HealthyAverage\"]  = df_longMicrobiome[\"Species\"].map(dict_speciesToHealthyAverage)\n",
    "df_longMicrobiome[\"HealthyMax\"]      = df_longMicrobiome[\"Species\"].map(dict_speciesToHealthyMax)\n",
    "\n",
    "df_longMicrobiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "20174398",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Complerness check of abundance values\n",
    "df_longMicrobiome['Abundance'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "1abe1d15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>Panel</th>\n",
       "      <th>DominantSpecies</th>\n",
       "      <th>HealthyAverage</th>\n",
       "      <th>HealthyMax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>p__Firmicutes;f__;g__</td>\n",
       "      <td>9.52</td>\n",
       "      <td>16S</td>\n",
       "      <td>Clostridium_sp_L2_50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>p__Firmicutes;f__;g__</td>\n",
       "      <td>9.52</td>\n",
       "      <td>16S</td>\n",
       "      <td>Clostridium_sp_L2_50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>p__Firmicutes;f__Lachnospiraceae;g__</td>\n",
       "      <td>4.83</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>p__Firmicutes;f__Lachnospiraceae;g__</td>\n",
       "      <td>4.83</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>4/3/12</td>\n",
       "      <td>p__Firmicutes;f__;g__</td>\n",
       "      <td>2.23</td>\n",
       "      <td>16S</td>\n",
       "      <td>Clostridium_sp_L2_50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16058</th>\n",
       "      <td>7/28/19</td>\n",
       "      <td>p__Firmicutes;f__Lachnospiraceae;g__</td>\n",
       "      <td>6.90</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16102</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Firmicutes;f__;g__</td>\n",
       "      <td>9.82</td>\n",
       "      <td>16S</td>\n",
       "      <td>Clostridium_sp_L2_50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16103</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Firmicutes;f__;g__</td>\n",
       "      <td>9.82</td>\n",
       "      <td>16S</td>\n",
       "      <td>Clostridium_sp_L2_50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16109</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Firmicutes;f__Lachnospiraceae;g__</td>\n",
       "      <td>9.44</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16110</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Firmicutes;f__Lachnospiraceae;g__</td>\n",
       "      <td>9.44</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>636 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Date                               Species  Abundance Panel  \\\n",
       "190    12/28/11                 p__Firmicutes;f__;g__       9.52   16S   \n",
       "191    12/28/11                 p__Firmicutes;f__;g__       9.52   16S   \n",
       "197    12/28/11  p__Firmicutes;f__Lachnospiraceae;g__       4.83   16S   \n",
       "198    12/28/11  p__Firmicutes;f__Lachnospiraceae;g__       4.83   16S   \n",
       "421      4/3/12                 p__Firmicutes;f__;g__       2.23   16S   \n",
       "...         ...                                   ...        ...   ...   \n",
       "16058   7/28/19  p__Firmicutes;f__Lachnospiraceae;g__       6.90   16S   \n",
       "16102   8/11/19                 p__Firmicutes;f__;g__       9.82   16S   \n",
       "16103   8/11/19                 p__Firmicutes;f__;g__       9.82   16S   \n",
       "16109   8/11/19  p__Firmicutes;f__Lachnospiraceae;g__       9.44   16S   \n",
       "16110   8/11/19  p__Firmicutes;f__Lachnospiraceae;g__       9.44   16S   \n",
       "\n",
       "            DominantSpecies  HealthyAverage  HealthyMax  \n",
       "190    Clostridium_sp_L2_50             NaN         NaN  \n",
       "191    Clostridium_sp_L2_50             NaN         NaN  \n",
       "197                    None             NaN         NaN  \n",
       "198                    None             NaN         NaN  \n",
       "421    Clostridium_sp_L2_50             NaN         NaN  \n",
       "...                     ...             ...         ...  \n",
       "16058                  None             NaN         NaN  \n",
       "16102  Clostridium_sp_L2_50             NaN         NaN  \n",
       "16103  Clostridium_sp_L2_50             NaN         NaN  \n",
       "16109                  None             NaN         NaN  \n",
       "16110                  None             NaN         NaN  \n",
       "\n",
       "[636 rows x 7 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Code to check for duplicates\n",
    "dups = (\n",
    "    df_longMicrobiome[df_longMicrobiome.duplicated([\"Date\",\"Species\"], keep=False)]\n",
    "    .sort_values([\"Date\",\"Species\"])\n",
    ")\n",
    "dups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "6e2170d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/nf/dp5pbtr55rn8mww76lsrkt800000gn/T/ipykernel_80014/3982303624.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  all_same = dups.groupby(['Date','Species'])['Abundance'].nunique(dropna=True).le(1).all()\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "dups = df_longMicrobiome[df_longMicrobiome.duplicated(['Date','Species'], keep=False)].copy()\n",
    "dups['Abundance'] = pd.to_numeric(dups['Abundance'], errors='coerce')\n",
    "\n",
    "all_same = dups.groupby(['Date','Species'])['Abundance'].nunique(dropna=True).le(1).all()\n",
    "print(all_same)  # True => all the same, False => at least one group differs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "4cadba62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removed 318 duplicate rows.\n"
     ]
    }
   ],
   "source": [
    "# remove duplicate (Date, Species) rows (values are identical)\n",
    "n_before = len(df_longMicrobiome)\n",
    "\n",
    "df_longMicrobiome = (\n",
    "    df_longMicrobiome\n",
    "        .sort_values([\"Date\", \"Species\"])\n",
    "        .drop_duplicates(subset=[\"Date\", \"Species\"], keep=\"first\")\n",
    "        .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "print(f\"Removed {n_before - len(df_longMicrobiome)} duplicate rows.\")\n",
    "# sanity check\n",
    "assert not df_longMicrobiome.duplicated([\"Date\",\"Species\"]).any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "a1cb9e78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>Panel</th>\n",
       "      <th>DominantSpecies</th>\n",
       "      <th>HealthyAverage</th>\n",
       "      <th>HealthyMax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Adlercreutzia_equolifaciens</td>\n",
       "      <td>0.031</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Aeromonas_unclassified</td>\n",
       "      <td>0.000</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Aggregatibacter_segnis</td>\n",
       "      <td>0.000</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Aggregatibacter_unclassified</td>\n",
       "      <td>0.000</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12/28/11</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>12.328</td>\n",
       "      <td>Xin Fang MetaphlAN 2 Metagenomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.984429</td>\n",
       "      <td>8.439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15820</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Proteobacteria;f__Pasteurellaceae;g__Aggreg...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15821</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Proteobacteria;f__Pseudomonadaceae;g__Pseud...</td>\n",
       "      <td>0.060</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15822</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Synergistetes;f__Synergistaceae;g__Synergistes</td>\n",
       "      <td>0.030</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15823</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Tenericutes;f__;g__</td>\n",
       "      <td>0.580</td>\n",
       "      <td>16S</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15824</th>\n",
       "      <td>8/11/19</td>\n",
       "      <td>p__Verrucomicrobia;f__Verrucomicrobiaceae;g__A...</td>\n",
       "      <td>1.600</td>\n",
       "      <td>16S</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15825 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Date                                            Species  Abundance  \\\n",
       "0      12/28/11                        Adlercreutzia_equolifaciens      0.031   \n",
       "1      12/28/11                             Aeromonas_unclassified      0.000   \n",
       "2      12/28/11                             Aggregatibacter_segnis      0.000   \n",
       "3      12/28/11                       Aggregatibacter_unclassified      0.000   \n",
       "4      12/28/11                            Akkermansia_muciniphila     12.328   \n",
       "...         ...                                                ...        ...   \n",
       "15820   8/11/19  p__Proteobacteria;f__Pasteurellaceae;g__Aggreg...      0.000   \n",
       "15821   8/11/19  p__Proteobacteria;f__Pseudomonadaceae;g__Pseud...      0.060   \n",
       "15822   8/11/19  p__Synergistetes;f__Synergistaceae;g__Synergistes      0.030   \n",
       "15823   8/11/19                             p__Tenericutes;f__;g__      0.580   \n",
       "15824   8/11/19  p__Verrucomicrobia;f__Verrucomicrobiaceae;g__A...      1.600   \n",
       "\n",
       "                                   Panel          DominantSpecies  \\\n",
       "0      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "1      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "2      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "3      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "4      Xin Fang MetaphlAN 2 Metagenomics                      NaN   \n",
       "...                                  ...                      ...   \n",
       "15820                                16S                     None   \n",
       "15821                                16S                     None   \n",
       "15822                                16S                     None   \n",
       "15823                                16S                     None   \n",
       "15824                                16S  Akkermansia_muciniphila   \n",
       "\n",
       "       HealthyAverage  HealthyMax  \n",
       "0                 NaN         NaN  \n",
       "1                 NaN         NaN  \n",
       "2                 NaN         NaN  \n",
       "3                 NaN         NaN  \n",
       "4            0.984429       8.439  \n",
       "...               ...         ...  \n",
       "15820             NaN         NaN  \n",
       "15821             NaN         NaN  \n",
       "15822             NaN         NaN  \n",
       "15823             NaN         NaN  \n",
       "15824             NaN         NaN  \n",
       "\n",
       "[15825 rows x 7 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_longMicrobiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "7f4b0e12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out the long format file\n",
    "df_longMicrobiome.to_csv(fp_microbiomeLong, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2752fc84",
   "metadata": {},
   "source": [
    "## Create Gut Microbiome Wide DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "07c0ad48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>12/28/11</th>\n",
       "      <th>4/3/12</th>\n",
       "      <th>8/7/12</th>\n",
       "      <th>11/6/12</th>\n",
       "      <th>1/26/13</th>\n",
       "      <th>2/8/13</th>\n",
       "      <th>2/5/13</th>\n",
       "      <th>3/24/13</th>\n",
       "      <th>4/7/13</th>\n",
       "      <th>...</th>\n",
       "      <th>4/14/19</th>\n",
       "      <th>4/28/19</th>\n",
       "      <th>5/12/19</th>\n",
       "      <th>5/19/19</th>\n",
       "      <th>6/2/19</th>\n",
       "      <th>6/16/19</th>\n",
       "      <th>6/23/19</th>\n",
       "      <th>7/14/19</th>\n",
       "      <th>7/28/19</th>\n",
       "      <th>8/11/19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adlercreutzia_equolifaciens</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.035</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.064</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.036</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.006</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aeromonas_unclassified</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aggregatibacter_segnis</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.562</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aggregatibacter_unclassified</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.137</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>12.328</td>\n",
       "      <td>0.950</td>\n",
       "      <td>3.842</td>\n",
       "      <td>5.579</td>\n",
       "      <td>3.518</td>\n",
       "      <td>3.212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.063</td>\n",
       "      <td>0.892</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>p__Proteobacteria;f__Pasteurellaceae;g__Aggreg...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.030</td>\n",
       "      <td>1.030</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>p__Proteobacteria;f__Pseudomonadaceae;g__Pseud...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.030</td>\n",
       "      <td>...</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>p__Synergistetes;f__Synergistaceae;g__Synergistes</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.020</td>\n",
       "      <td>...</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>p__Tenericutes;f__;g__</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.160</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.160</td>\n",
       "      <td>...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>p__Verrucomicrobia;f__Verrucomicrobiaceae;g__A...</td>\n",
       "      <td>14.770</td>\n",
       "      <td>0.850</td>\n",
       "      <td>4.130</td>\n",
       "      <td>5.670</td>\n",
       "      <td>3.120</td>\n",
       "      <td>2.460</td>\n",
       "      <td>6.58</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.490</td>\n",
       "      <td>...</td>\n",
       "      <td>2.88</td>\n",
       "      <td>4.43</td>\n",
       "      <td>1.29</td>\n",
       "      <td>3.73</td>\n",
       "      <td>3.96</td>\n",
       "      <td>1.19</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.66</td>\n",
       "      <td>5.50</td>\n",
       "      <td>1.60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>229 rows × 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Date                                            Species  12/28/11  4/3/12  \\\n",
       "0                           Adlercreutzia_equolifaciens     0.031   0.035   \n",
       "1                                Aeromonas_unclassified     0.000   0.000   \n",
       "2                                Aggregatibacter_segnis     0.000   0.000   \n",
       "3                          Aggregatibacter_unclassified     0.000   0.000   \n",
       "4                               Akkermansia_muciniphila    12.328   0.950   \n",
       "..                                                  ...       ...     ...   \n",
       "224   p__Proteobacteria;f__Pasteurellaceae;g__Aggreg...     0.000   0.000   \n",
       "225   p__Proteobacteria;f__Pseudomonadaceae;g__Pseud...     0.000   0.010   \n",
       "226   p__Synergistetes;f__Synergistaceae;g__Synergistes     0.100   0.000   \n",
       "227                              p__Tenericutes;f__;g__     0.990   0.000   \n",
       "228   p__Verrucomicrobia;f__Verrucomicrobiaceae;g__A...    14.770   0.850   \n",
       "\n",
       "Date  8/7/12  11/6/12  1/26/13  2/8/13  2/5/13  3/24/13  4/7/13  ...  4/14/19  \\\n",
       "0      0.143    0.064    0.041   0.036     NaN    0.008   0.006  ...      NaN   \n",
       "1      0.000    0.000    0.000   0.000     NaN    0.000   0.000  ...      NaN   \n",
       "2      0.000    0.000    0.000   0.014     NaN    0.015   0.562  ...      NaN   \n",
       "3      0.000    0.000    0.000   0.000     NaN    0.000   0.137  ...      NaN   \n",
       "4      3.842    5.579    3.518   3.212     NaN    1.063   0.892  ...      NaN   \n",
       "..       ...      ...      ...     ...     ...      ...     ...  ...      ...   \n",
       "224    0.000    0.000    0.000   0.040    0.29    0.030   1.030  ...     0.00   \n",
       "225    0.000    0.010    0.010   0.010    0.01    0.010   0.030  ...     0.02   \n",
       "226    0.020    0.050    0.110   0.040    0.12    0.010   0.020  ...     0.07   \n",
       "227    0.010    0.030    0.160   0.110    0.05    0.350   0.160  ...     0.64   \n",
       "228    4.130    5.670    3.120   2.460    6.58    0.490   0.490  ...     2.88   \n",
       "\n",
       "Date  4/28/19  5/12/19  5/19/19  6/2/19  6/16/19  6/23/19  7/14/19  7/28/19  \\\n",
       "0         NaN      NaN      NaN     NaN      NaN      NaN      NaN      NaN   \n",
       "1         NaN      NaN      NaN     NaN      NaN      NaN      NaN      NaN   \n",
       "2         NaN      NaN      NaN     NaN      NaN      NaN      NaN      NaN   \n",
       "3         NaN      NaN      NaN     NaN      NaN      NaN      NaN      NaN   \n",
       "4         NaN      NaN      NaN     NaN      NaN      NaN      NaN      NaN   \n",
       "..        ...      ...      ...     ...      ...      ...      ...      ...   \n",
       "224      0.00     0.00     0.00    0.00     0.00     0.00     0.00     0.00   \n",
       "225      0.00     0.10     0.10    0.33     0.09     0.00     0.09     0.08   \n",
       "226      0.03     0.06     0.01    0.00     0.00     0.03     0.05     0.11   \n",
       "227      0.50     0.22     0.33    0.39     0.39     0.40     0.27     0.16   \n",
       "228      4.43     1.29     3.73    3.96     1.19     0.97     0.66     5.50   \n",
       "\n",
       "Date  8/11/19  \n",
       "0         NaN  \n",
       "1         NaN  \n",
       "2         NaN  \n",
       "3         NaN  \n",
       "4         NaN  \n",
       "..        ...  \n",
       "224      0.00  \n",
       "225      0.06  \n",
       "226      0.03  \n",
       "227      0.58  \n",
       "228      1.60  \n",
       "\n",
       "[229 rows x 160 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create wide format by pivoting\n",
    "df_wideMicrobiome = df_longMicrobiome.pivot(index=\"Species\", columns=\"Date\", values=\"Abundance\").reset_index()\n",
    "df_wideMicrobiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "af4e9129",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the wide format file\n",
    "df_wideMicrobiome.to_csv(fp_microbiomeWide, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "basic",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
